---
title: "GASPARONI dataset"
author: "Lanyu Zhang, Tiago C. Silva, Lily Wang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    theme: lumen
    highlight: null
    toc: true
    number_sections: true
    df_print: paged
    code_download: true
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data retrival

```{R, message = FALSE, results = 'hide'}
library(dplyr)
library(minfi)
```

```{R}
cohort <- "GASPARONI"
data.dir <- file.path("../DATASETS",cohort) 
data.dir.table <- "../DATASETS/Summary_Table/" 
data.dir.raw <- "../../coMethDMR_metaAnalysis/code_validation/Meta_analysis_code/DATASETS/GASPARONI/step2_read_minfi/"
data.dir.bsfilter <- file.path(data.dir,"step2_bsConvFilter/") 
data.dir.clinical.filter <- file.path(data.dir,"step3_clinical_available_filtering/") 
data.dir.probes.qc <- file.path(data.dir,"step4_probesQC_filtering/") 
data.dir.probes.normalization <- file.path(data.dir,"step5_normalization/") 
data.dir.pca <- file.path(data.dir,"step6_pca_filtering/") 
data.dir.neuron <- file.path(data.dir,"step7_neuron_comp/") 
data.dir.single.cpg.pval <- file.path(data.dir,"step8_single_cpg_pval/") 
data.dir.residuals <- file.path(data.dir,"step9_residuals/") 
data.dir.median <- file.path(data.dir,"step10_median/") 
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Confirm sex status sing DNA methylation values

```{R}
load(file = paste0(data.dir.raw, "Gasparoni.rda"))

# Create a MethylSet object from RGSet
MSet <- preprocessRaw(RGSet)

# Create [Genomic]MethylSet object
GMset <- mapToGenome(MSet)

# Get predicted sex status
estSex <- getSex(GMset)

# Compare predicted gender with phenotype gender
realSex <- data.frame(
  sample = paste(phenoData$geo_accession, phenoData$sentrix_id.ch1, phenoData$sentrix_position.ch1, sep = "_"),
  realSex = phenoData$Sex.ch1,
  stringsAsFactors = FALSE
)
compareSex <- merge(
  as.data.frame(estSex), realSex,
  by.x = "row.names",
  by.y = "sample")

identical(compareSex$predictedSex, compareSex$realSex)

print(compareSex)
```

# Data QC 

## Bilsufite conversion filtering

Removing samples with bisulfiteConversion lower than 88.

```{R qc_samples_packages, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
#library(sesame)
#library(sesameData)
```


```{R}
load(file = paste0(data.dir.raw, "Gasparoni.rda"))
bs <- data.frame(bisulfiteConversion = bscon(RGSet))
bsFilteredOut <- row.names(bs)[bs$bisulfiteConversion < 88]
nb.samples <- ncol(RGSet)
RGSet <- RGSet[,!colnames(RGSet) %in% bsFilteredOut]
nb.samples.bc.filtered <-  ncol(RGSet)
phenoData <- phenoData[match(substr(colnames(RGSet),1,10), phenoData$geo_accession),]
```

```{R, eval = FALSE}
save(RGSet,
     nb.samples,
     bs,
     phenoData,
     nb.samples.bc.filtered, 
     file = paste0(data.dir.bsfilter, "/RGSet_bsfiltered.rda"))
```

```{R, include = FALSE, eval = TRUE}
load(file = paste0(data.dir.bsfilter, "/RGSet_bsfiltered.rda"))
```

```{R, include = FALSE, eval=FALSE}
ggpubr::gghistogram(bs$bisulfiteConversion,xlab = "bisulfite Conversion" )
```

## Clinical data filtering

```{R}
dim(phenoData)

phenoData$braak_stage.ch1 <- phenoData$braak_stage.ch1 %>% as.numeric()
phenoData$age.ch1 <- phenoData$age.ch1 %>% as.numeric()
### Subset rows and columns
pheno_df <- phenoData  %>% as.data.frame() %>%
  dplyr::filter(
    source_name_ch1 == "Frontal Cortex" &
      !is.na(phenoData$braak_stage.ch1) &
      phenoData$characteristics_ch1 == "cell type: bulk"
  ) %>% dplyr::select(
    c(
      "geo_accession",
      "donor_id.ch1",
      "sentrix_id.ch1",
      "age.ch1",
      "Sex.ch1",
      "braak_stage.ch1"
    )
  )

### Rename vars
colnames(pheno_df) <- c(
  "sample", "subject.id", "slide", "age.brain", "sex", "stage"
)

dim(pheno_df) 
nb.samples.with.clinical <- nrow(pheno_df)
## phenotype dataset
save(RGSet,
     nb.samples.with.clinical,
     pheno_df,
     file = paste0(data.dir.clinical.filter, "/gasparoni_bs_and_clinical_filtered.rda"))
```

## Probes QC

Input: 

- RGSet.RDS
- beta_mat.RDS

Output: 

- beta_CG_XY_SNPfiltered.RDS


```{R}
##### 1. subset to probes with detection P <= 0.01 #############################
### Read in RGSet and beta_mat
load(paste0(data.dir.clinical.filter, "/gasparoni_bs_and_clinical_filtered.rda"))
```

```{R, message = FALSE, results = 'hide'}
### subset to probes with detection P <= 0.01
library(minfi)
```

```{R}
nb.probes <- nrow(RGSet)
detP <- detectionP(RGSet, type = "m+u")
failed.01 <- detP > 0.01
passedProbes <- rownames(failed.01)[rowMeans(failed.01) == 0] 

beta_mat <- getBeta(RGSet) 

beta_mat <- beta_mat[passedProbes, ]
nb.probes.detectP <- nrow(beta_mat)

##### 2. keep only probes that start with "cg" #################################
beta_mat <- beta_mat[grep("cg",rownames(beta_mat)),]
dim(beta_mat) 
nb.probes.detectP.cg <- nrow(beta_mat)

##### 3. drop probes where SNP with MAF >= 0.01 in the last 5 bp of the probe ##
```

```{R, message = FALSE, warning = FALSE}
library(DMRcate)
beta_mat <- rmSNPandCH(
  object = beta_mat,
  dist = 5, 
  mafcut = 0.01, 
  and = TRUE,
  rmcrosshyb = FALSE,
  rmXY = FALSE
) 
nb.probes.cg.dmrcate <- nrow(beta_mat)

##### 4. drop probes in chrM ###################################################
library(sesame)

probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

beta_mat <- beta_mat[
  row.names(beta_mat) %in% row.names(probes.info)[!probes.info$seqnames %in% "chrM"],
]
nb.probes.cg.dmrcate.chrM <- nrow(beta_mat)

##### 5. Output datasets #######################################################
save(
  pheno_df,
  nb.probes.detectP,
  nb.probes.detectP.cg,
  nb.probes.cg.dmrcate,
  nb.probes.cg.dmrcate.chrM,
  beta_mat,
  file = paste0(data.dir.probes.qc, "beta_CG_XY_SNPfiltered_mat.rda")
)
```

# Normalization

- Quantile normalization and BMIQ normalization

Input: 

- beta_CG_XY_SNPfiltered_mat.RDS
- RGSet.RDS
- pheno_df.RDS
- full.annot.RDS

Output: 

- bs.csv
- pheno_df.RDS
- QNBMIQ.RDS

```{R qcSamplesData}
load(paste0(data.dir.probes.qc, "beta_CG_XY_SNPfiltered_mat.rda"))
```


## Quantile normalization

### Check if sex chrom have higher/lower intensities than autosomes

#### Create function to extract separate matrix for each chromosome

```{R, message = FALSE, results = 'hide'}
library(sesame)
library(ggplot2)
```

```{R}
probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

findBetaChr <- function(data, chrom){
  data %>%
    as.data.frame() %>% # has to turn matrix into df for next step
    tibble::rownames_to_column() %>% # turn rownames to a column, so rownames won't be deleted after filtering rows in next step
    filter(rowname %in% row.names(probes.info[probes.info$seqnames == chrom,])) %>%
    tibble::column_to_rownames() %>%
    as.matrix()
}
```

#### Compare beta values between gender

##### on chrX

```{R}
betaChrX <- findBetaChr(data = beta_mat, chrom = "chrX")
betaChrX_long <- data.frame(
  beta = as.vector(betaChrX),
  sample = rep(substr(colnames(betaChrX), 1,10), each = nrow(betaChrX)),
  stringsAsFactors = FALSE
)
betaChrX_long <- merge(
  betaChrX_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)
  
ggplot(betaChrX_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start=1, end=0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome X") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

##### on chrY

```{R}
betaChrY <- findBetaChr(data = beta_mat, chrom = "chrY")
betaChrY_long <- data.frame(
  beta = as.vector(betaChrY),
  sample = rep(substr(colnames(betaChrY), 1,10), each = nrow(betaChrY)),
  stringsAsFactors = FALSE
)
betaChrY_long <- merge(
  betaChrY_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)

ggplot(betaChrY_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start=1, end=0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome Y") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

#### Boxplot of methylation by chromosomes  

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)

chrSex <- paste0("chr", c("X", "Y"))
betaChrSex_ls <- lapply(seq_along(chrSex), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrSex[i])
  data.frame(beta = as.vector(dat), chrom = chrSex[i], stringsAsFactors = FALSE)
})
betaChrSex_df <- do.call(rbind, betaChrSex_ls)

betaChr_df <- rbind(betaChrAuto_df, betaChrSex_df)

# # No enough memory to load all the data points and plot the figure, so use function boxplot instead
# ggplot(betaChr_df,
#        aes(x = chrom, y = beta, fill = chrom)) +
#   stat_boxplot(geom ='errorbar', width = 0.4, linetype = 1) +
#   geom_boxplot(width = 0.4, alpha = 1, outlier.shape = 1, outlier.size = 2) +
#   scale_fill_grey(start=1, end=0.6) +
#   labs(x = "chromosome type", y = "DNA methylation beta values",
#        title = "DNA methylation level vs. chromosome type") +
#   theme_bw()

boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")

pdf(file = paste0(data.dir.probes.normalization, "beta_vs_chrom.pdf"))
boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")
dev.off()
```

#### Boxplot of methylation by gender on autosomes, X, Y

```{R, message = FALSE, results = 'hide'}
library(quantro)
```

##### on autosomes

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){findBetaChr(data = beta_mat, chrom = chrAuto[i])})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)
matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrX 

```{R}
betaChrX_df <- findBetaChr(data = beta_mat, chrom = "chrX")
matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrY

```{R}
betaChrY_df <- findBetaChr(data = beta_mat,chrom = "chrY")
matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### save plots

```{R}
pdf (
  paste0(data.dir.probes.normalization, "boxPlotBeforeNormalization.pdf")
)

matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

dev.off()
```

### normalization by group

```{R, message = FALSE, results = 'hide'}
library(lumi)
```

#### split beta_mat into sub-matrics by females autosomes, females X, males autosomes, males X, males Y

```{R}
### split matrix by sex first
beta_mat_female <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "F"]]
beta_mat_male <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "M"]]

### split female beta matrix by chromosome (auto, X)
chrAuto <- paste0("chr", 1:22)
beta_mat_female_auto_ls <- lapply(seq_along(chrAuto), function(i){
  findBetaChr(data = beta_mat_female, chrom = chrAuto[i])})
beta_mat_female_auto <- do.call(rbind, beta_mat_female_auto_ls)

beta_mat_female_X <- findBetaChr(data = beta_mat_female, chrom = "chrX")

### split male beta matrix by chromosome (auto, X, Y)
chrAuto <- paste0("chr", 1:22)
beta_mat_male_auto_ls <- lapply(seq_along(chrAuto), function(i){
  findBetaChr(data = beta_mat_male, chrom = chrAuto[i])})
beta_mat_male_auto <- do.call(rbind, beta_mat_male_auto_ls)

beta_mat_male_X <- findBetaChr(data = beta_mat_male, chrom = "chrX")

beta_mat_male_Y <- findBetaChr(data = beta_mat_male, chrom = "chrY")
```

#### normalization 

```{R}
### female autosomes
betaQN_female_auto <- lumiN(x.lumi = beta_mat_female_auto, method = "quantile")
dim(betaQN_female_auto)

### female chromosome X
betaQN_female_X <- lumiN(x.lumi = beta_mat_female_X, method = "quantile")
dim(betaQN_female_X)

### male autosomes
betaQN_male_auto <- lumiN(x.lumi = beta_mat_male_auto, method = "quantile")
dim(betaQN_male_auto)

### male chromosome X
betaQN_male_X <- lumiN(x.lumi = beta_mat_male_X, method = "quantile")
dim(betaQN_male_X)

### male chromosome Y
betaQN_male_Y <- lumiN(x.lumi = beta_mat_male_Y, method = "quantile")
dim(betaQN_male_Y)
```

## BMIQ

```{R, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
library(sesame)
library(sesameData)
library(sm)
```

### test percentage of type I and type II probes then normalize the matrix

#### on autosomes

##### for females

```{R}
### Order annotation in the same order as beta matrix
annotType <- sesameDataGet("HM450.hg19.manifest")
annotType$designTypeNumeric <- ifelse(annotType$designType == "I",1,2)

### Density plot for type I and type II probes
betaQNCompleteCol1_female_auto <- betaQN_female_auto[complete.cases(betaQN_female_auto[,1]), ]
annotTypeCompleteCol1_female_auto <- annotType[row.names(betaQNCompleteCol1_female_auto), ]

sm.density.compare(
    betaQNCompleteCol1_female_auto[,1],
    annotTypeCompleteCol1_female_auto$designTypeNumeric) +
  title(main = "Density plot for type I and type II probes on female autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_auto),names(annotType))]
table(type12)
```
```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_female_auto <- plyr::aaply(
  betaQN_female_auto, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_female_auto) <- substr(colnames(betaQN_BMIQ_female_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_auto <- betaQN_male_auto[complete.cases(betaQN_male_auto[,1]), ]
annotTypeCompleteCol1_male_auto <- annotType[row.names(betaQNCompleteCol1_male_auto), ]

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_auto),names(annotType))]
table(type12)
```
```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_male_auto <- plyr::aaply(
  betaQN_male_auto, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_male_auto) <- substr(colnames(betaQN_BMIQ_male_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

#### on chromosome X

##### for females

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_female_X <- betaQN_female_X[complete.cases(betaQN_female_X[,1]), ]
annotTypeCompleteCol1_female_X <- annotType[row.names(betaQNCompleteCol1_female_X), ]

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_X),names(annotType))]
table(type12)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_X <- betaQN_male_X[complete.cases(betaQN_male_X[,1]), ]
annotTypeCompleteCol1_male_X <- annotType[row.names(betaQNCompleteCol1_male_X), ]

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_X),names(annotType))]
table(type12)
```

#### on chromosome Y

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_Y <- betaQN_male_Y[complete.cases(betaQN_male_Y[,1]), ]
annotTypeCompleteCol1_male_Y <- annotType[row.names(betaQNCompleteCol1_male_Y), ]

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_Y),names(annotType))]
table(type12)
```

### save density plots and normalized datasets

```{R}
### save plots
pdf(paste0(data.dir.probes.normalization, "densityPlotByProbeType.pdf"))

sm.density.compare(
  betaQNCompleteCol1_female_auto[,1],
  annotTypeCompleteCol1_female_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female autosomes")

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

dev.off()

### combined normalized matrices
betaQN_BMIQ_female <- rbind(betaQN_BMIQ_female_auto, betaQN_female_X)
betaQN_BMIQ_male <- rbind(betaQN_BMIQ_male_auto, betaQN_male_X, betaQN_male_Y)
save(betaQN_BMIQ_female, betaQN_BMIQ_male, pheno_df, file = paste0(data.dir.probes.normalization, "GASPARONI_QNBMIQ.rda"))

```

# Outliers detection - PCA analysis

- Select most variable probes and perform PCA analysis

Input: 

- GASPARONI_QNBMIQ.rds  
- pheno_df.RDS  

Output: 

- GASPARONI_PCs_usingBetas.csv, 
- PCA plots
- GASPARONI_QNBMIQ_PCfiltered.RDS
- pheno_df.RDS

```{R}
# plotPCA and OrderDataBySd functions
devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")
```
```{R}
### merge in group info
## add center and scale
## compare M values vs. beta values
## with and without center / scale
load(paste0(data.dir.probes.normalization, "GASPARONI_QNBMIQ.rda"))
```

## for females  

```{R}
pheno_female_df <- pheno_df %>% filter(sex == "F")
pheno_female_df$stage3 <- ifelse(
  pheno_female_df$stage %in% c(0,1,2), "0-2",
  ifelse(pheno_female_df$stage %in% c(3,4), "3-4", "5-6")
)
betaQN_BMIQ_female <- betaQN_BMIQ_female[ , pheno_female_df$sample]

### transform to m values
mvalue_mat <- log2(betaQN_BMIQ_female / (1 - betaQN_BMIQ_female)) #dim: 457402 30

### order matrix by most variable probes on top
betaOrd_mat <- OrderDataBySd(betaQN_BMIQ_female) #dim: 457402 30
mOrd_mat <- OrderDataBySd(mvalue_mat)  #dim: 457402 30

identical(pheno_female_df$sample, colnames(betaOrd_mat))
identical(pheno_female_df$sample, colnames(mOrd_mat))

pca <- prcomp(t(betaOrd_mat[1:50000, ]),
              center = TRUE,
              scale = TRUE)

d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])

meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)

meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)

out3sdPC1_1 <- meanPC1 - 3 * sdPC1
out3sdPC1_2 <- meanPC1 + 3 * sdPC1

out3sdPC2_1 <- meanPC2 - 3 * sdPC2
out3sdPC2_2 <- meanPC2 + 3 * sdPC2

d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1

d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1

readr::write_csv(d, paste0(data.dir.pca, "GASPARONI_PCs_usingBetas_female.csv"))
```

```{R, eval = FALSE, include = FALSE}
### 2.PCA plot
library(ggplot2)
library(ggrepel)

### Beata values
byStage <- plotPCA(
  dataset = "Gasparoni: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_female_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

### M values
byStage <- plotPCA(
  dataset = "Gasparoni: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_female_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```

```{R}
### Filter samples by PCA, save files
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered_female <- betaQN_BMIQ_female[, rownames(noOutliers)] #dim: 457402 29
saveRDS(betaQN_BMIQ_PCfiltered_female, paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered_female.RDS"))

pheno_female_df <- pheno_female_df[pheno_female_df$sample %in% rownames(noOutliers),] #dim: 29 7
saveRDS(pheno_female_df, paste0(data.dir.pca, "pheno_female_df.RDS"))
```

## for males  

```{R}
pheno_male_df <- pheno_df %>% filter(sex == "M")
pheno_male_df$stage3 <- ifelse(
  pheno_male_df$stage %in% c(0,1,2), "0-2",
  ifelse(pheno_male_df$stage %in% c(3,4), "3-4", "5-6")
)
betaQN_BMIQ_male <- betaQN_BMIQ_male[ , pheno_male_df$sample]

### transform to m values
mvalue_mat <- log2(betaQN_BMIQ_male / (1 - betaQN_BMIQ_male)) #dim: 457464 27

### order matrix by most variable probes on top
betaOrd_mat <- OrderDataBySd(betaQN_BMIQ_male) #dim: 457464 27
mOrd_mat <- OrderDataBySd(mvalue_mat)  #dim: 457464 27

identical(pheno_male_df$sample, colnames(betaOrd_mat))
identical(pheno_male_df$sample, colnames(mOrd_mat))

pca <- prcomp(t(betaOrd_mat[1:50000, ]),
              center = TRUE,
              scale = TRUE)

d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])

meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)

meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)

out3sdPC1_1 <- meanPC1 - 3 * sdPC1
out3sdPC1_2 <- meanPC1 + 3 * sdPC1

out3sdPC2_1 <- meanPC2 - 3 * sdPC2
out3sdPC2_2 <- meanPC2 + 3 * sdPC2

d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1

d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1

readr::write_csv(d, paste0(data.dir.pca, "GASPARONI_PCs_usingBetas_male.csv"))
```

```{R, eval = FALSE, include = FALSE}
### 2.PCA plot
library(ggplot2)
library(ggrepel)

### Beata values
byStage <- plotPCA(
  dataset = "Gasparoni: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_male_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

### M values
byStage <- plotPCA(
  dataset = "Gasparoni: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_male_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```


```{R}
### Filter samples by PCA, save files
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered_male <- betaQN_BMIQ_male[, rownames(noOutliers)] #dim: 457464 27
saveRDS(betaQN_BMIQ_PCfiltered_male, paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered_male.RDS"))

pheno_male_df <- pheno_male_df[pheno_male_df$sample %in% rownames(noOutliers),] #dim: 27 7
saveRDS(pheno_male_df, paste0(data.dir.pca, "pheno_male_df.RDS"))
```

# Summary after QC steps

## Data and metadata
```{R}
nb.samples.with.clinical.after.pca.female <- ncol(betaQN_BMIQ_PCfiltered_female)
nb.samples.with.clinical.after.pca.male <- ncol(betaQN_BMIQ_PCfiltered_male)
dim(betaQN_BMIQ_PCfiltered_female)
dim(betaQN_BMIQ_PCfiltered_male)
dim(pheno_female_df)
dim(pheno_male_df)

pheno_female_df %>% 
  DT::datatable(filter = 'top',
                style = "bootstrap",
                extensions = 'Buttons',
                options = list(scrollX = TRUE, 
                               dom = 'Bfrtip',
                               buttons = I('colvis'),
                               keys = TRUE, 
                               pageLength = 10), 
                rownames = FALSE,
                caption = "Female samples metadata")

pheno_male_df %>% 
  DT::datatable(filter = 'top',
                style = "bootstrap",
                extensions = 'Buttons',
                options = list(scrollX = TRUE, 
                               dom = 'Bfrtip',
                               buttons = I('colvis'),
                               keys = TRUE, 
                               pageLength = 10), 
                rownames = FALSE,
                caption = "Male samples metadata")
```

```{R, eval = FALSE, include = FALSE}
ggpubr::gghistogram(data = pheno_female_df, x = "stage",bins = 8)
# ggpubr::gghistogram(data = pheno_female_df, x = "stage",bins = 8,facet.by = "sex",fill = "sex")
ggpubr::gghistogram(data = pheno_female_df, x = "age.brain",bins = 20)
# ggpubr::gghistogram(data = pheno_female_df, x = "age.brain",bins = 20,fill = "sex",facet.by = "sex")

ggpubr::gghistogram(data = pheno_male_df, x = "stage",bins = 8)
# ggpubr::gghistogram(data = pheno_male_df, x = "stage",bins = 8,facet.by = "sex",fill = "sex")
ggpubr::gghistogram(data = pheno_male_df, x = "age.brain",bins = 20)
# ggpubr::gghistogram(data = pheno_male_df, x = "age.brain",bins = 20,fill = "sex",facet.by = "sex")
```


## Numbers of samples and probes removed in each step

```{R}
df.samples <- data.frame("Number of samples" =  c(nb.samples, 
                                                  nb.samples.bc.filtered,
                                                  nb.samples.with.clinical, 
                                                  nb.samples.with.clinical.after.pca.female + nb.samples.with.clinical.after.pca.male),
                         "Description" = c("total number of samples",
                                           "samples with bisulfate conversion > 88",
                                           "samples with clinical data",
                                           "Samples after PCA"),
                         "Difference" = c("-",
                                          nb.samples.bc.filtered - nb.samples ,
                                          nb.samples.with.clinical - nb.samples.bc.filtered ,
                                          nb.samples.with.clinical.after.pca.female + nb.samples.with.clinical.after.pca.male - nb.samples.with.clinical)
)    
df.samples           

# Create summary table
df.probes <- data.frame("Number of probes" = c(nb.probes,
                                               nb.probes.detectP, 
                                               nb.probes.detectP.cg,
                                               nb.probes.cg.dmrcate,
                                               nb.probes.cg.dmrcate.chrM),
                        "Description" = c("total number of probes in raw data",
                                          "detection P < 0.01",
                                          "only probes that start with cg",
                                          "DMRcate",
                                          "after deleting probes on chrM"),
                        "Difference" = c("-",
                                         nb.probes.detectP - nb.probes ,
                                         nb.probes.detectP.cg - nb.probes.detectP,
                                         nb.probes.cg.dmrcate - nb.probes.detectP.cg,
                                         nb.probes.cg.dmrcate.chrM - nb.probes.cg.dmrcate)
)
df.probes

save(df.samples,df.probes,file = file.path(data.dir.table, "GASPARONI_table.rda"))
```

<!-- # Compute neuron proportion -->

<!-- Data from  https://www.tandfonline.com/doi/full/10.4161/epi.23924 -->

<!-- - Input: Gasparoni_QNBMIQ_PCfiltered.RDS, pheno_df.RDS -->

<!-- - Output: pheno_withNeuronProp_df.RDS -->

<!-- ```{R} -->
<!-- objects <- load("../../CET/CETS_Image.RData") -->
<!-- objects -->
<!-- ``` -->

<!-- ## Get reference profile from Caucasions + controls  -->
<!-- ```{R} -->
<!-- idx <- list( -->
<!--   controlNeuron = pdBrain$celltype == "N" & pdBrain$diag == "Control" & pdBrain$ethnicity == "Caucasian", -->
<!--   controlGlia   = pdBrain$celltype == "G" & pdBrain$diag == "Control" & pdBrain$ethnicity == "Caucasian" -->
<!-- ) -->

<!-- refProfile <- getReference(brain, idx) -->


<!-- ##### 2. Estimate proportions of neurons in PFC samples ######################## -->

<!-- ### Limit to 10,000 cpgs in the refProfile dataset -->
<!-- pfc <- readRDS(paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered.RDS")) #dim: 433656 59 -->

<!-- selected <- rownames(pfc) %in% rownames(refProfile) -->

<!-- pfc.refcpgs <- pfc[selected, ]  -->

<!-- ### Estimate proportion of neurons -->
<!-- prop <- data.frame(estProportion(pfc.refcpgs, profile = refProfile)) -->
<!-- colnames(prop) <- "prop.neuron" -->

<!-- ##### 3. Merge pfc.refcpgs with phenotype file ################################# -->
<!-- pheno <- readRDS(paste0(data.dir.pca, "pheno_df.RDS")) -->

<!-- pheno_final <- merge( -->
<!--   pheno, -->
<!--   prop, -->
<!--   by.x = "sample", -->
<!--   by.y = "row.names" -->
<!-- ) -->

<!-- saveRDS(pheno_final, paste0(data.dir.neuron, "pheno_withNeuronProp_df.RDS")) -->
<!-- ``` -->

<!-- # Linear regression by cpgs Methylation  -->

<!-- Input:  -->

<!-- - Gasparoni_QNBMIQ_PCfiltered.RDS, -->
<!-- - pheno59_withNeuronProp_df.RDS -->

<!-- Output: -->

<!-- - Gasparoni_single_cpg_pVal_df.csv -->

<!-- ## Import datasets -->

<!-- ```{R} -->
<!-- beta_mat <- readRDS(paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered.RDS"))  -->
<!-- pheno_df <- readRDS(paste0(data.dir.neuron, "pheno_withNeuronProp_df.RDS"))  -->
<!-- ``` -->

<!-- ## Test all cpgs -->

<!-- ```{R, eval = TRUE} -->
<!-- ### Compute M values -->
<!-- mval_mat <- log2(beta_mat / (1 - beta_mat)) -->

<!-- pheno_df$Sample <- pheno_df$sample -->

<!-- identical(pheno_df$Sample, colnames(mval_mat)) -->

<!-- pheno_df$sex <- as.factor(pheno_df$sex) -->
<!-- pheno_df$slide <- as.factor(pheno_df$slide) -->
<!-- # If rosmap cohort, don't forget batch effect -->

<!-- str(pheno_df) -->

<!-- is(pheno_df$stage,"numeric") -->
<!-- is(pheno_df$age.brain,"numeric") -->
<!-- is(pheno_df$prop.neuron,"numeric") -->
<!-- ``` -->


<!-- ```{R, eval = FALSE} -->
<!-- predictors_char <- "stage" -->
<!-- covariates_char <- c("age.brain", "sex", "prop.neuron", "slide") -->


<!-- doParallel::registerDoParallel(cores = parallel::detectCores()/2) -->
<!-- devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e") -->

<!-- results_ordered_df <- plyr::adply(mval_mat,1, function(row){ -->

<!--   sumOneRegion_df <- data.frame(t(row)) -->

<!--   result <- TestSingleRegion( -->
<!--     predictors_char = predictors_char, -->
<!--     covariates_char = covariates_char, -->
<!--     pheno_df = pheno_df, -->
<!--     sumOneRegion_df = sumOneRegion_df -->
<!--   ) -->
<!--   result -->
<!-- }, .progress = "time",.parallel = TRUE,.id = "cpg") -->
<!-- colnames(results_ordered_df)[1] <- "cpg" -->

<!-- identical(row.names(mval_mat), results_ordered_df$cpg %>% as.character()) -->

<!-- results_ordered_df$fdr <- p.adjust( -->
<!--     results_ordered_df$pValue, -->
<!--     method = "fdr" -->
<!-- ) -->

<!-- write.csv( -->
<!--   results_ordered_df, -->
<!--   paste0(data.dir.single.cpg.pval, "Gasparoni_single_cpg_pVal_df.csv"), -->
<!--   row.names = FALSE -->
<!-- ) -->
<!-- ``` -->

<!-- ```{R} -->
<!-- results_ordered_df <- readr::read_csv( -->
<!--   paste0(data.dir.single.cpg.pval, "Gasparoni_single_cpg_pVal_df.csv"), -->
<!--   col_types = readr::cols()) -->
<!-- results_ordered_df -->
<!-- ``` -->


<!-- # Linear regression by regions median Methylation  -->

<!-- ## Residuals control and coMethylated Regions -->

<!-- 1. Take residuals -->
<!-- 2. Find co-methylated regions -->

<!-- Input:  -->

<!-- - QNBMIQ_PCfiltered -->
<!-- - pheno_withNeuronProp_df -->

<!-- Output:  -->

<!-- - QNBMIQ_PCfiltered_mvalResiduals -->
<!-- - residuals_cometh_ls -->

<!-- ### Residuals -->

<!-- ```{R, eval = FALSE} -->
<!-- ##### 1. Import datasets ####################################################### -->
<!-- beta_mat <- readRDS(paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered.RDS")) #dim:433656 59 -->
<!-- pheno_df <- readRDS(paste0(data.dir.neuron, "pheno_withNeuronProp_df.RDS")) #dim:59 7 -->

<!-- ### Compute M values -->
<!-- mvalue_mat <- log2(beta_mat / (1 - beta_mat)) -->

<!-- ### Reorder samples based on pheno_df -->
<!-- mvalue_mat <- mvalue_mat[, pheno_df$sample] -->

<!-- identical(colnames(mvalue_mat),  pheno_df$sample) -->

<!-- ### Take residuals -->
<!-- lmF <- function(mval){ -->
<!--   fitE <- lm( -->
<!--     as.numeric(mval) ~ age.brain + sex + prop.neuron + as.character(slide), #add batch if rosmap -->
<!--     data = pheno_df, -->
<!--     na.action = na.exclude -->
<!--   ) -->
<!--   residuals (fitE) -->
<!-- } -->
<!-- doParallel::registerDoParallel(cores = 4) -->
<!-- resid <- plyr::adply(mvalue_mat,1,.fun = lmF,.progress = "time",.parallel = TRUE) -->
<!-- rownames(resid) <- resid[,1] -->
<!-- resid[,1] <- NULL -->
<!-- colnames(resid) <- colnames(mvalue_mat) -->
<!-- dim(resid) -->
<!-- dim(mvalue_mat) -->

<!-- ### Save dataset -->
<!-- saveRDS( -->
<!--   resid, -->
<!--   paste0(data.dir.residuals, "Gasparoni_QNBMIQ_PCfiltered_mvalResiduals.RDS") -->
<!-- ) -->
<!-- ``` -->

<!-- ### Find co-methylated regions -->

<!-- ```{R, eval = FALSE} -->
<!-- ### Import datasets -->
<!-- mvalue_residuals_mat <- readRDS( -->
<!--   paste0(data.dir.residuals, "Gasparoni_QNBMIQ_PCfiltered_mvalResiduals.RDS") -->
<!-- ) -->

<!-- ### Call in functions -->
<!-- library(coMethDMR) -->
<!-- library(BiocParallel) -->

<!-- probes.cluster.all <- coMethDMR::getPredefinedCluster(arrayType = "450k", -->
<!--                                                       clusterType = "regions") -->

<!-- ncores <- parallel::detectCores()/2 -->
<!-- ### Find co-methylated clusters -->
<!-- coMeth_ls <- CoMethAllRegions( -->
<!--   dnam = mvalue_residuals_mat,       -->
<!--   betaToM = FALSE,                    -->
<!--   CpGs_ls = probes.cluster.all, -->
<!--   arrayType = "450k", -->
<!--   rDropThresh_num = 0.4, -->
<!--   minPairwiseCorr = NULL, -->
<!--   method = "spearman",              -->
<!--   returnAllCpGs = TRUE,               -->
<!--   output = "all", -->
<!--   nCores_int = ncores, -->
<!--   progressbar = FALSE -->
<!-- ) -->

<!-- saveRDS( -->
<!--   coMeth_ls, -->
<!--   paste0(data.dir.residuals,"Gasparoni_residuals_cometh_input_ls.RDS") -->
<!-- ) -->
<!-- ``` -->

<!-- ## Linear regression by regions median Methylation  -->

<!-- 1. Calculate medians by cluster and sample -->
<!-- 2. linear regression -->

<!-- Input:  -->

<!-- - QNBMIQ_PCfiltered, -->
<!-- - pheno_withNeuronProp_df -->
<!-- - residuals_cometh_input_ls -->

<!-- Output:  -->

<!-- - info_df -->
<!-- - mediansMval_df -->
<!-- - linear_df_aging -->

<!-- ### Calculate medians by cluster and sample -->

<!-- ```{R, eval = FALSE} -->
<!-- ### Import datasets -->
<!-- beta_mat <- readRDS(paste0(data.dir.pca, -->
<!--                            "Gasparoni_QNBMIQ_PCfiltered.RDS")) -->
<!-- pheno_df <- readRDS(paste0(data.dir.neuron, "pheno_withNeuronProp_df.RDS")) -->
<!-- mval_mat <- log2(beta_mat / (1 - beta_mat)) %>% as.matrix() -->
<!-- coMeth_ls <- readRDS( -->
<!--   paste0(data.dir.residuals, "Gasparoni_residuals_cometh_input_ls.RDS") -->
<!-- ) -->

<!-- ### Create info dataset -->
<!-- input_cometh <- data.frame( -->
<!--   inputRegion = coMeth_ls$inputRegion_chr, -->
<!--   nCoMethRegion = coMeth_ls$nCoMethRegions_num, -->
<!--   coMethRegion = names(coMeth_ls$coMeth_ls), -->
<!--   nCpGs = unlist(lapply(coMeth_ls$coMeth_ls, length), use.names = FALSE), -->
<!--   stringsAsFactors = FALSE -->
<!-- ) -->

<!-- input_cometh_nodup <- input_cometh[ -->
<!--   !duplicated(input_cometh$coMethRegion), -->
<!--   ] -->
<!-- colnames(input_cometh_nodup) <- c( -->
<!--   paste0(cohort, "_inputRegion"), -->
<!--   paste0(cohort, "_nCoMethRegion"), -->
<!--   paste0(cohort, "_coMethRegion"), -->
<!--   paste0(cohort, "_nCpGs") -->
<!-- ) -->

<!-- saveRDS( -->
<!--   input_cometh_nodup, -->
<!--   paste0(data.dir.median, cohort,"_info_df.rds") -->
<!-- ) -->

<!-- ### Take median of probes in each cluster for each sample -->
<!-- filename <-  paste0(paste0(data.dir.median, cohort, "_mediansMval_df.rds")) -->
<!-- library(robustbase) -->
<!-- mval_mat <- mval_mat[rownames(mval_mat) %in% unlist(coMeth_ls$coMeth_ls),] -->
<!-- if(!file.exists(filename)){ -->
<!--   medianMval.df <- plyr::ldply( -->
<!--     coMeth_ls$coMeth_ls[!duplicated(names(coMeth_ls$coMeth_ls))], -->
<!--     function(probes){ -->
<!--       colMedians(mval_mat[as.character(probes),], na.rm = TRUE) -->
<!--     }, -->
<!--     .progress = "time" -->
<!--   ) -->
<!--   medianMval.df$.id <- NULL -->
<!--   colnames(medianMval.df) <- colnames(mval_mat) -->
<!--   saveRDS(medianMval.df, file = filename) -->
<!-- } else { -->
<!--   medianMval.df <- readRDS(filename) -->
<!-- } -->
<!-- ``` -->

<!-- ### Test all regions -- linear regressions -->

<!-- ```{R, eval = TRUE} -->
<!-- ### Import datasets -->
<!-- cohort <- "Gasparoni" -->
<!-- info_df <- readRDS(dir(data.dir.median, pattern = "info", full.names = TRUE)) -->
<!-- mediansMval_df <- readRDS(dir(data.dir.median, pattern = "mediansMval", full.names = TRUE)) -->
<!-- pheno_df <- readRDS(paste0(data.dir.neuron, "pheno_withNeuronProp_df.RDS"))  -->

<!-- ### Check variables before fitting model -->
<!-- pheno_df$Sample <- pheno_df$sample -->
<!-- identical(pheno_df$Sample, colnames(mediansMval_df)) -->

<!-- pheno_df$sex <- as.factor(pheno_df$sex) -->
<!-- pheno_df$slide <- as.factor(pheno_df$slide) -->
<!-- # If rosmap cohort, don't forget batch effect -->

<!-- str(pheno_df) -->
<!-- ``` -->

<!-- ```{R, eval = FALSE} -->
<!-- devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e") -->

<!-- predictors_char <- "stage" -->
<!-- covariates_char <- c("age.brain", "sex", "prop.neuron", "slide") -->

<!-- res_df <- TestAllRegions_noInfo( -->
<!--   predictors_char = predictors_char, -->
<!--   covariates_char = covariates_char, -->
<!--   pheno_df = pheno_df, -->
<!--   summarizedRegions_df = mediansMval_df -->
<!-- ) -->

<!-- colnames(res_df) <- c( -->
<!--   paste0(cohort, "_estimate"), -->
<!--   paste0(cohort, "_se"), -->
<!--   paste0(cohort, "_pVal"), -->
<!--   paste0(cohort, "_fdr") -->
<!-- ) -->

<!-- res_withInfo_df <- cbind(info_df, res_df) -->

<!-- saveRDS( -->
<!--   res_withInfo_df, -->
<!--   paste0(data.dir.median,  cohort, "_linear_df.rds") -->
<!-- ) -->
<!-- ``` -->

<!-- ```{R} -->
<!-- file <- dir(data.dir.median,pattern = paste0(".*linear_df"), -->
<!--             recursive = T, -->
<!--             full.names = TRUE, -->
<!--             ignore.case = T) -->
<!-- file -->
<!-- res_withInfo_df <- readRDS(file) -->
<!-- dim(res_withInfo_df) -->
<!-- res_withInfo_df -->
<!-- ``` -->

<!-- # Data final -->

<!-- ```{R} -->
<!-- dir(path = data.dir,recursive = T, pattern = ".rda|.csv|.RDS") -->
<!-- ``` -->

<!-- # Session information -->
<!-- ```{R} -->
<!-- devtools::session_info() -->
<!-- ``` -->
