---
title: "LONDON dataset"
author: "Lanyu Zhang, Tiago C. Silva, Lily Wang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    theme: lumen
    highlight: kate
    toc: true
    number_sections: true
    df_print: paged
    code_download: true
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Data retrival

```{R, message = FALSE, results = 'hide'}
library(dplyr)
library(minfi)
library(IlluminaHumanMethylation450kmanifest)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```

```{R}
cohort <- "LONDON"
data.dir <- file.path("../DATASETS",cohort) 
data.dir.table <- "../DATASETS/Summary_Table/" 
data.dir.raw <- "../../coMethDMR_metaAnalysis/code_validation/Meta_analysis_code/DATASETS/LONDON/step1_download/"
data.dir.paper1 <- "../../coMethDMR_metaAnalysis/code_validation/Meta_analysis_code/DATASETS/LONDON/step3_probesQC_filtering/"
data.dir.clinical.filter <- file.path(data.dir,"step2_clinical_available_filtering/") 
data.dir.probes.qc <- file.path(data.dir,"step3_probesQC_filtering/") 
data.dir.probes.normalization <- file.path(data.dir,"step4_normalization/") 
data.dir.pca <- file.path(data.dir,"step5_pca_filtering/") 
data.dir.neuron <- file.path(data.dir,"step6_neuron_comp/") 
data.dir.single.cpg.pval <- file.path(data.dir,"step7_single_cpg_pval/") 
data.dir.residuals <- file.path(data.dir,"step8_residuals/") 
data.dir.median <- file.path(data.dir,"step9_median/") 
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Confirm sex status

```{R, include = FALSE, eval=TRUE}
pheno_df <- readRDS(paste0(data.dir.raw, "GSE59685_pheno.rds"))
beta_mat <- readRDS(paste0(data.dir.paper1, "beta110_PFC_CG_XY_SNPfiltered_mat.rds"))
```

because London cohort does not have idat files, we are using hcluster instead of minfi::getSex here

```{R, message = FALSE, results = 'hide'}
library(amap)
library(dplyr)
library(ggplot2)
library(dendextend)
library(ggdendro)
library(magrittr)
```

```{r}
pheno_df <- pheno_df[pheno_df$geo_accession %in% colnames(beta_mat), ]
beta_mat <- beta_mat[,pheno_df$geo_accession]
colnames(beta_mat) <- ifelse(pheno_df$Sex.ch1 == "FEMALE", "female", "male")

beta_mat_t <- t(beta_mat)
dend <- beta_mat_t %>%
  Dist(nbproc = 20) %>%
  hclust %>%
  as.dendrogram
dend %>% set("labels_cex", 0.7) %>% plot
# label color by group
# https://cran.r-project.org/web/packages/dendextend/vignettes/FAQ.html#how-to-color-a-dendrograms-labels-according-to-defined-groups-in-r

```

# Data QC

Description:

- Subset Files:
  1. prefrontal cortex tissue and stage != "Exclude" (n = 110)
  2. variables -- sample, subject_id, sentrix_id, slide, age, sex, stage

- subset methylation data to match subjects in pheno data (n = 110)

 Input: GSE59685_pheno.csv, GSE59685_assay.RDS
 
 Output: pheno_PFC_df.RDS, beta110_PFC_mat.RDS

## Clinical data filtering

```{R, include = FALSE, eval=TRUE}
pheno_df <- readRDS(paste0(data.dir.raw, "GSE59685_pheno.rds"))
beta_mat <- readRDS(paste0(data.dir.raw, "GSE59685_assay.rds"))
nb.samples <- nrow(pheno_df)
nb.female.samples <- sum(pheno_df$Sex.ch1 == "FEMALE")
nb.male.samples <- sum(pheno_df$Sex.ch1 == "MALE")
```

```{R, eval = TRUE}
phenoPFC_df <- pheno_df %>%
  as.data.frame %>%
  dplyr::filter(source_name_ch1 == "frontal cortex" &
                  braak.stage.ch1 != "Exclude") %>%
  dplyr::select(
    c(
      "geo_accession",
      "subjectid.ch1",
      "barcode.ch1",
      "age.brain.ch1",
      "Sex.ch1",
      "braak.stage.ch1"
    )
  )

### Rename vars
colnames(phenoPFC_df) <- c("sample",
                           "subject.id",
                           "sentrix_id",
                           "age.brain",
                           "sex",
                           "stage")

### Get slide from sentrix_id
  # e.g. "6042316048_R05C01"(sentrix_id) -- "6042316048"(slide) and "R05C01"(array)
phenoPFC_df$slide <- gsub("_[[:alnum:]]*$","",phenoPFC_df$sentrix_id)
phenoPFC_df$stage <- as.numeric(phenoPFC_df$stage)
phenoPFC_df$age.brain <- as.numeric(phenoPFC_df$age.brain)

### Order final pheno_df
pheno_df <- phenoPFC_df[, c(
  "sample", "subject.id", "sentrix_id", "slide", "age.brain", "sex", "stage"
)] #dim 110 7

### Subset methylation data
beta_mat <- beta_mat[, match(pheno_df$sample,colnames(beta_mat))]

nb.samples.with.clinical <- nrow(pheno_df)
nb.female.samples.with.clinical <- sum(pheno_df$sex == "FEMALE")
nb.male.samples.with.clinical <- sum(pheno_df$sex == "MALE")

save(beta_mat,
     nb.samples,
     nb.female.samples,
     nb.male.samples,
     nb.samples.with.clinical,
     nb.female.samples.with.clinical,
     nb.male.samples.with.clinical,
     pheno_df,
     file = paste0(data.dir.clinical.filter, "/london_clinical_filtered.rda"))
```

## Probes QC

 1. keep only probes that start with "cg"
 2. drop probes that are on X/Y
 3. drop probes where SNP with MAF >= 0.01 was present in the last 5 bp of the probe.

 Input: beta110_PFC_mat.RDS
 
 Output: beta110_PFC_CG_XY_SNPfiltered_mat.RDS

```{R, include=FALSE, eval=TRUE}
load(paste0(data.dir.clinical.filter, "/london_clinical_filtered.rda"))
```

```{R, message = FALSE, results = 'hide'}
library(minfi)
library(DMRcate)
library(sesame)
```

```{R}
### Find which chromosome each probe is on
probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

nb.probes <- nrow(probes.info)
nb.chrAuto.probes <- sum(probes.info$seqnames %in% paste0("chr", 1:22))
nb.chrX.probes <- sum(probes.info$seqnames == "chrX")
nb.chrY.probes <- sum(probes.info$seqnames == "chrY")
nb.chrM.probes <- sum(probes.info$seqnames == "chrM")
```

```{R}
### keep on probes with start with "cg"
beta_mat <- beta_mat[grep("cg",rownames(beta_mat)),]
probes.info <- probes.info[row.names(probes.info) %in% row.names(beta_mat),]

nb.probes.cg <- nrow(probes.info)
nb.chrAuto.probes.cg <- sum(probes.info$seqnames %in% paste0("chr", 1:22))
nb.chrX.probes.cg <- sum(probes.info$seqnames == "chrX")
nb.chrY.probes.cg <- sum(probes.info$seqnames == "chrY")
nb.chrM.probes.cg <- sum(probes.info$seqnames == "chrM")
```

```{R, message = FALSE, warning = FALSE}
### drop probes where SNP with MAF >= 0.01 in the last 5 bp of the probe
beta_mat <- rmSNPandCH(
  object = beta_mat,
  dist = 5,
  mafcut = 0.01,
  and = TRUE,
  rmcrosshyb = FALSE,
  rmXY = FALSE
)
probes.info <- probes.info[row.names(probes.info) %in% row.names(beta_mat),]

nb.probes.cg.dmrcate <- nrow(probes.info)
nb.chrAuto.probes.cg.dmrcate <- sum(probes.info$seqnames %in% paste0("chr", 1:22))
nb.chrX.probes.cg.dmrcate <- sum(probes.info$seqnames == "chrX")
nb.chrY.probes.cg.dmrcate <- sum(probes.info$seqnames == "chrY")
nb.chrM.probes.cg.dmrcate <- sum(probes.info$seqnames == "chrM")

### drop probes in chrM
probes.info <- probes.info[probes.info$seqnames != "chrM",]
beta_mat <- beta_mat[
  row.names(beta_mat) %in% row.names(probes.info),
]

nb.probes.dmrcate.chrM <- nrow(probes.info)
nb.chrAuto.probes.dmrcate.chrM <- sum(probes.info$seqnames %in% paste0("chr", 1:22))
nb.chrX.probes.dmrcate.chrM <- sum(probes.info$seqnames == "chrX")
nb.chrY.probes.dmrcate.chrM <- sum(probes.info$seqnames == "chrY")

save(
  pheno_df,
  beta_mat,
  nb.probes.cg,
  nb.chrAuto.probes.cg,
  nb.chrX.probes.cg,
  nb.chrY.probes.cg,
  nb.chrM.probes.cg,
  nb.probes.cg.dmrcate,
  nb.chrAuto.probes.cg.dmrcate,
  nb.chrX.probes.cg.dmrcate,
  nb.chrY.probes.cg.dmrcate,
  nb.chrM.probes.cg.dmrcate,
  nb.probes.dmrcate.chrM,
  nb.chrAuto.probes.dmrcate.chrM,
  nb.chrX.probes.dmrcate.chrM,
  nb.chrY.probes.dmrcate.chrM,
  file = paste0(data.dir.probes.qc, "/beta_CG_XY_SNPfiltered_mat.rda")
)
```

# Normalization

- Quantile normalization and BMIQ normalization

 Input: 
 
 - beta110_PFC_CG_XY_SNPfiltered_mat.RDS
 - pheno_PFC_df.RDS
 - full.annot.RDS
 
 Output: 
 
- London_PFC_QNBMIQ.RDS


```{R qcSamplesData, include = FALSE, eval=TRUE}
load(paste0(data.dir.probes.qc, "/beta_CG_XY_SNPfiltered_mat.rda"))
```

## Quantile normalization

### Check if sex chrom have higher/lower intensities than autosomes

#### Create function to extract separate matrix for each chromosome

```{R, message = FALSE, results = 'hide'}
library(sesame)
library(ggplot2)
```

```{R}
probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

findBetaChr <- function(data, chrom){
  data %>%
    as.data.frame() %>% # has to turn matrix into df for next step
    tibble::rownames_to_column() %>% # turn rownames to a column, so rownames won't be deleted after filtering rows in next step
    filter(rowname %in% row.names(probes.info[probes.info$seqnames == chrom,])) %>%
    tibble::column_to_rownames() %>%
    as.matrix()
}
```

#### Compare beta values between gender

##### on chrX

```{R}
betaChrX <- findBetaChr(data = beta_mat, chrom = "chrX")
betaChrX_long <- data.frame(
  beta = as.vector(betaChrX),
  sample = rep(substr(colnames(betaChrX), 1,10), each = nrow(betaChrX)),
  stringsAsFactors = FALSE
)
betaChrX_long <- merge(
  betaChrX_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)
  
ggplot(betaChrX_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start=1, end=0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome X") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

##### on chrY

```{R}
betaChrY <- findBetaChr(data = beta_mat, chrom = "chrY")
betaChrY_long <- data.frame(
  beta = as.vector(betaChrY),
  sample = rep(substr(colnames(betaChrY), 1,10), each = nrow(betaChrY)),
  stringsAsFactors = FALSE
)
betaChrY_long <- merge(
  betaChrY_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)

ggplot(betaChrY_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start=1, end=0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome Y") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

#### Boxplot of methylation by chromosomes  

##### overall  

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)

chrSex <- paste0("chr", c("X", "Y"))
betaChrSex_ls <- lapply(seq_along(chrSex), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrSex[i])
  data.frame(beta = as.vector(dat), chrom = chrSex[i], stringsAsFactors = FALSE)
})
betaChrSex_df <- do.call(rbind, betaChrSex_ls)

betaChr_df <- rbind(betaChrAuto_df, betaChrSex_df)

# # No enough memory to load all the data points and plot the figure, so use function boxplot instead
# ggplot(betaChr_df,
#        aes(x = chrom, y = beta, fill = chrom)) +
#   stat_boxplot(geom ='errorbar', width = 0.4, linetype = 1) +
#   geom_boxplot(width = 0.4, alpha = 1, outlier.shape = 1, outlier.size = 2) +
#   scale_fill_grey(start=1, end=0.6) +
#   labs(x = "chromosome type", y = "DNA methylation beta values",
#        title = "DNA methylation level vs. chromosome type") +
#   theme_bw()

boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")

pdf(file = paste0(data.dir.probes.normalization, "/beta_vs_chrom.pdf"))
boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")
dev.off()
```

##### by gender  

```{R}
### split matrix by sex first
beta_mat_female <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "FEMALE"]]
beta_mat_male <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "MALE"]]

### for females
chrAuto <- paste0("chr", 1:22)
female_auto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat_female, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
female_auto_df <- do.call(rbind, female_auto_ls)

female_sex_df <- findBetaChr(data = beta_mat_female, chrom = "chrX")
female_sex_df <- data.frame(beta = as.vector(female_sex_df), chrom = "chromosome X", stringsAsFactors = FALSE)

female_df <- rbind(female_auto_df, female_sex_df)

boxplot(
  beta ~ chrom, data = female_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for females")

### for males
chrAuto <- paste0("chr", 1:22)
male_auto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat_male, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
male_auto_df <- do.call(rbind, male_auto_ls)

male_X_df <- findBetaChr(data = beta_mat_male, chrom = "chrX")
male_X_df <- data.frame(beta = as.vector(male_X_df), chrom = "chromosome X", stringsAsFactors = FALSE)

male_Y_df <- findBetaChr(data = beta_mat_male, chrom = "chrY")
male_Y_df <- data.frame(beta = as.vector(male_Y_df), chrom = "chromosome Y", stringsAsFactors = FALSE)

male_df <- rbind(male_auto_df, male_X_df, male_Y_df)

boxplot(
  beta ~ chrom, data = male_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for males")

### save plots
pdf(file = paste0(data.dir.probes.normalization, "/beta_vs_chrom_by_gender.pdf"))

boxplot(
  beta ~ chrom, data = female_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for females")

boxplot(
  beta ~ chrom, data = male_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for males")

dev.off()
```

#### Boxplot of methylation by gender on autosomes, X, Y

```{R, message = FALSE, results = 'hide'}
library(quantro)
```

##### on autosomes

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){findBetaChr(data = beta_mat, chrom = chrAuto[i])})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)
matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrX 

```{R}
betaChrX_df <- findBetaChr(data = beta_mat, chrom = "chrX")
matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrY

```{R}
betaChrY_df <- findBetaChr(data = beta_mat,chrom = "chrY")
matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### save plots

```{R}
pdf (
  paste0(data.dir.probes.normalization, "/boxPlotBeforeNormalization.pdf")
)

matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

dev.off()
```

### normalization by group

```{R, message = FALSE, results = 'hide'}
library(lumi)
```

#### split beta_mat into sub-matrics by females autosomes, females X, males autosomes, males X, males Y

```{R}
### split matrix by sex first
beta_mat_female <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "FEMALE"]]
beta_mat_male <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "MALE"]]

### split female beta matrix by chromosome (auto, X)
chrAuto <- paste0("chr", 1:22)
beta_mat_female_auto_ls <- lapply(seq_along(chrAuto), function(i){
  findBetaChr(data = beta_mat_female, chrom = chrAuto[i])})
beta_mat_female_auto <- do.call(rbind, beta_mat_female_auto_ls)

beta_mat_female_X <- findBetaChr(data = beta_mat_female, chrom = "chrX")

### split male beta matrix by chromosome (auto, X, Y)
chrAuto <- paste0("chr", 1:22)
beta_mat_male_auto_ls <- lapply(seq_along(chrAuto), function(i){
  findBetaChr(data = beta_mat_male, chrom = chrAuto[i])})
beta_mat_male_auto <- do.call(rbind, beta_mat_male_auto_ls)

beta_mat_male_X <- findBetaChr(data = beta_mat_male, chrom = "chrX")

beta_mat_male_Y <- findBetaChr(data = beta_mat_male, chrom = "chrY")
```

#### normalization 

```{R}
### female autosomes
betaQN_female_auto <- lumiN(x.lumi = beta_mat_female_auto, method = "quantile")
dim(betaQN_female_auto)

### female chromosome X
betaQN_female_X <- lumiN(x.lumi = beta_mat_female_X, method = "quantile")
dim(betaQN_female_X)

### male autosomes
betaQN_male_auto <- lumiN(x.lumi = beta_mat_male_auto, method = "quantile")
dim(betaQN_male_auto)

### male chromosome X
betaQN_male_X <- lumiN(x.lumi = beta_mat_male_X, method = "quantile")
dim(betaQN_male_X)

### male chromosome Y
betaQN_male_Y <- lumiN(x.lumi = beta_mat_male_Y, method = "quantile")
dim(betaQN_male_Y)
```

##### Female plots after normalization 

```{R}
pheno_female_df <- pheno_df %>% filter(sex == "FEMALE")
matboxplot(betaQN_female_auto,
           groupFactor = pheno_female_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_female_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_female_X,
           groupFactor = pheno_female_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_female_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

```

##### Male plots after normalization 

```{R}
pheno_male_df <- pheno_df %>% filter(sex == "MALE")
matboxplot(betaQN_male_auto,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_male_X,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_male_Y,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

## BMIQ

```{R, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
library(sesame)
library(sesameData)
library(sm)
```

### test percentage of type I and type II probes then normalize the matrix

#### on autosomes

##### for females

```{R}
### Order annotation in the same order as beta matrix
annotType <- sesameDataGet("HM450.hg19.manifest")
annotType$designTypeNumeric <- ifelse(annotType$designType == "I",1,2)

### Density plot for type I and type II probes
betaQNCompleteCol1_female_auto <- betaQN_female_auto[complete.cases(betaQN_female_auto[,1]), ]
annotTypeCompleteCol1_female_auto <- annotType[row.names(betaQNCompleteCol1_female_auto), ]

sm.density.compare(
    betaQNCompleteCol1_female_auto[,1],
    annotTypeCompleteCol1_female_auto$designTypeNumeric) +
  title(main = "Density plot for type I and type II probes on female autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_auto),names(annotType))]
table(type12)
```

```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_female_auto <- plyr::aaply(
  betaQN_female_auto, 2,
  function(x){
    norm_ls <- wateRmelon::BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_female_auto) <- substr(colnames(betaQN_BMIQ_female_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_auto <- betaQN_male_auto[complete.cases(betaQN_male_auto[,1]), ]
annotTypeCompleteCol1_male_auto <- annotType[row.names(betaQNCompleteCol1_male_auto), ]

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_auto),names(annotType))]
table(type12)
```
```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_male_auto <- plyr::aaply(
  betaQN_male_auto, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_male_auto) <- substr(colnames(betaQN_BMIQ_male_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

#### on chromosome X

##### for females

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_female_X <- betaQN_female_X[complete.cases(betaQN_female_X[,1]), ]
annotTypeCompleteCol1_female_X <- annotType[row.names(betaQNCompleteCol1_female_X), ]

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_X),names(annotType))]
table(type12)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_X <- betaQN_male_X[complete.cases(betaQN_male_X[,1]), ]
annotTypeCompleteCol1_male_X <- annotType[row.names(betaQNCompleteCol1_male_X), ]

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_X),names(annotType))]
table(type12)
```

#### on chromosome Y

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_Y <- betaQN_male_Y[complete.cases(betaQN_male_Y[,1]), ]
annotTypeCompleteCol1_male_Y <- annotType[row.names(betaQNCompleteCol1_male_Y), ]

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_Y),names(annotType))]
table(type12)
```

### save density plots and normalized datasets

```{R}
### save plots
pdf(paste0(data.dir.probes.normalization, "/densityPlotByProbeType.pdf"))

sm.density.compare(
  betaQNCompleteCol1_female_auto[,1],
  annotTypeCompleteCol1_female_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female autosomes")

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

dev.off()

### combined normalized matrices
betaQN_BMIQ_female <- rbind(betaQN_BMIQ_female_auto, betaQN_female_X)
betaQN_BMIQ_male <- rbind(betaQN_BMIQ_male_auto, betaQN_male_X, betaQN_male_Y)
save(betaQN_BMIQ_female, betaQN_BMIQ_male, pheno_df, file = paste0(data.dir.probes.normalization, "/LONDON_QNBMIQ.rda"))

```

# Outliers detection - PCA analysis


Description: 

1. estimate standard deviation for each probe
2. select most variable probes (e.g. n = 50,000)
3. pca analysis
4. filter outliers

Input: 

- QNBMIQ.rds
- pheno_df.RDS

Output: 

- PCs_usingBetas.csv
- PCA plots
- QNBMIQ_PCfiltered.RDS
- pheno_df.RDS


```{R}
# plotPCA and OrderDataBySd functions
devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")

beta_mat <- readRDS(paste0(data.dir.probes.normalization, "London_PFC_QNBMIQ.RDS")) #dim: 437713 110

pheno_df <- readRDS(paste0(data.dir.clinical.filter, "pheno_PFC_df.RDS")) #dim: 110 7

identical(colnames(beta_mat), pheno_df$sample)

### transform to m values
mvalue_mat <- log2(beta_mat/(1 - beta_mat)) #dim: 437713 110

pheno_df <- subset(pheno_df, pheno_df$sample %in% colnames(beta_mat)) #dim: 110 7

pheno_df$stage3 <- pheno_df$stage
pheno_df$stage3[pheno_df$stage <= 2] <- '0-2'
pheno_df$stage3[pheno_df$stage > 2 & pheno_df$stage < 5] <- '3-4'
pheno_df$stage3[pheno_df$stage >= 5] <- '5-6'
table(pheno_df$stage3)

##### 1.Order matrix by most variable probes on top ############################

betaOrd_mat <- OrderDataBySd(beta_mat) #dim: 437713 110

mOrd_mat <- OrderDataBySd(mvalue_mat)  #dim: 437713 110

betaOrd_matPlot <- betaOrd_mat[, pheno_df$sample] #dim: 437713 110
mOrd_matPlot <- mOrd_mat[, pheno_df$sample]       #dim: 437713 110
identical(pheno_df$sample, colnames(betaOrd_matPlot))
identical(pheno_df$sample, colnames(mOrd_matPlot))

expSorted_mat = betaOrd_mat #dim: 437713 110

pca <- prcomp(
  t(expSorted_mat[1:50000,]),
  center = TRUE,
  scale = TRUE
)


d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])

meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)

meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)

out3sdPC1_1 <- meanPC1 - 3*sdPC1
out3sdPC1_2 <- meanPC1 + 3*sdPC1

out3sdPC2_1 <- meanPC2 - 3*sdPC2
out3sdPC2_2 <- meanPC2 + 3*sdPC2

d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1

d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1

write.csv(d, paste0(data.dir.pca, "London_PFC_PCs_usingBetas.csv"))
```


```{R, include = FALSE, eval = FALSE}

##### 2.PCA plot ###############################################################
library(ggplot2)
library(ggrepel)

### beta values
byStage <- plotPCA(
  dataset = "London PFC: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

bySex <- plotPCA(
  dataset = "London PFC: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_df,
  group_char = "sex",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

### M values
byStage <- plotPCA(
  dataset = "London PFC: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

bySex <- plotPCA(
  dataset = "London PFC: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_df,
  group_char = "sex",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```

## Filter samples by PCA 
```{R, eval = FALSE}
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered <- beta_mat[, rownames(noOutliers)] #dim: 437713 107
saveRDS(betaQN_BMIQ_PCfiltered, paste0(data.dir.pca, "London_PFC_QNBMIQ_PCfiltered.RDS"))
saveRDS(betaQN_BMIQ_PCfiltered, paste0(data.dir.pca, "London_PFC_QNBMIQ_PCfiltered.RDS"))

pheno107_df <- pheno_df[pheno_df$sample %in% rownames(noOutliers),] #dim: 107 7
saveRDS(pheno107_df, paste0(data.dir.pca, "pheno107_PFC_df.RDS"))
saveRDS(pheno107_df, paste0(data.dir.pca, "pheno107_PFC_df.RDS"))
```

# Summary after QC steps

## Data and metadata
```{R}
betaQN_BMIQ_PCfiltered <- readRDS(paste0(data.dir.pca, "London_PFC_QNBMIQ_PCfiltered.RDS")) #dim: 433656 59
nb.samples.with.clinical.after.pca <- ncol(betaQN_BMIQ_PCfiltered)
pheno107_df <- readRDS(paste0(data.dir.pca, "pheno107_PFC_df.RDS"))
dim(betaQN_BMIQ_PCfiltered)
dim(pheno107_df)
pheno107_df %>% 
  DT::datatable(filter = 'top',
                style = "bootstrap",
                extensions = 'Buttons',
                options = list(scrollX = TRUE, 
                               dom = 'Bfrtip',
                               buttons = I('colvis'),
                               keys = TRUE, 
                               pageLength = 10), 
                rownames = FALSE,
                caption = "Samples metadata")
```

## Numbers of samples and probes removed in each step


```{R}
df.samples <- data.frame("Number of samples" =  c(nb.samples, 
                                                  nb.samples.with.clinical, 
                                                  nb.samples.with.clinical.after.pca),
                         "Description" = c("total number of samples",
                                           "samples with clinical data",
                                           "Samples after PCA"),
                         "Difference" = c("-",
                                          nb.samples.with.clinical - nb.samples ,
                                          nb.samples.with.clinical.after.pca - nb.samples.with.clinical)
)    
df.samples                     
# Create summary table
df.probes <- data.frame("Number of probes" = c(nb.probes,
                                               nb.probes.cg, 
                                               nb.probes.cg.dmrcate),
                        "Description" = c("total number of probes in raw data",
                                          "only probes that start with cg",
                                          "DMRcate"),
                        "Difference" = c("-",
                                         nb.probes.cg - nb.probes ,
                                         nb.probes.cg.dmrcate - nb.probes.cg)
)
df.probes
save(df.samples,df.probes,file = file.path(data.dir.table, "LONDON_table.rda"))
```

# Compute neuron proportion


Data from  https://www.tandfonline.com/doi/full/10.4161/epi.23924

- Input: London_PFC_QNBMIQ_PCfiltered.RDS, pheno107_PFC_df.RDS
- Output: pheno107_PFC_withNeuronProp_df.RDS

```{R}
objects <- load("../../CET/CETS_Image.RData")
objects
```

## Get reference profile from Caucasions + controls 
```{R}
idx <- list(
  controlNeuron = pdBrain$celltype == "N" & pdBrain$diag == "Control" & pdBrain$ethnicity == "Caucasian",
  controlGlia   = pdBrain$celltype == "G" & pdBrain$diag == "Control" & pdBrain$ethnicity == "Caucasian"
)

refProfile <- getReference(brain, idx)
head(refProfile)


##### 2. Estimate proportions of neurons in PFC samples ########################

### Limit to 10,000 cpgs in the refProfile dataset
pfc <- readRDS(paste0(data.dir.pca, "London_PFC_QNBMIQ_PCfiltered.RDS")) #dim: 433656 59

selected <- rownames(pfc) %in% rownames(refProfile)

pfc.refcpgs <- pfc[selected, ] #dim: 9530 59

### Estimate proportion of neurons
prop <- data.frame(estProportion(pfc.refcpgs, profile = refProfile))
colnames(prop) <- "prop.neuron"


##### 3. Merge pfc.refcpgs with phenotype file #################################
pheno <- readRDS(paste0(data.dir.pca, "pheno107_PFC_df.RDS"))

pheno_final <- merge(
  pheno,
  prop,
  by.x = "sample",
  by.y = "row.names"
)

saveRDS(pheno_final, paste0(data.dir.neuron, "pheno107_PFC_withNeuronProp_df.RDS"))
```

# Linear regression by cpgs Methylation 

Input: 

  - London_PFC_QNBMIQ_PCfiltered.RDS,
  - pheno107_PFC_withNeuronProp_df.RDS

Output:

  - London_PFC_single_cpg_pVal_df.csv

## Import datasets

```{R}
beta_mat <- readRDS(paste0(data.dir.pca, "London_PFC_QNBMIQ_PCfiltered.RDS")) #dim: 437713 107
pheno_df <- readRDS(paste0(data.dir.neuron, "pheno107_PFC_withNeuronProp_df.RDS")) #dim:107 8
```

## Test all regions

```{R, eval = TRUE}
### Compute M values
mval_mat <- log2(beta_mat / (1 - beta_mat))

pheno_df$Sample <- pheno_df$sample

identical(pheno_df$Sample, colnames(mval_mat))

pheno_df$sex <- as.factor(pheno_df$sex)
pheno_df$slide <- as.factor(pheno_df$slide)
# If rosmap cohort, don't forget batch effect

is(pheno_df$stage,"numeric")
is(pheno_df$age.brain,"numeric")
is(pheno_df$prop.neuron,"numeric")

str(pheno_df)
```

```{R, eval = FALSE}

predictors_char <- "stage"
covariates_char <- c("age.brain", "sex", "prop.neuron", "slide")


doParallel::registerDoParallel(cores = parallel::detectCores()/2)
devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")

results_ordered_df <- plyr::adply(mval_mat,1, function(row){
  
  sumOneRegion_df <- data.frame(t(row))

  result <- TestSingleRegion(
    predictors_char = predictors_char,
    covariates_char = covariates_char,
    pheno_df = pheno_df,
    sumOneRegion_df = sumOneRegion_df
  )
  result
}, .progress = "time",.parallel = TRUE,.id = "cpg")
colnames(results_ordered_df)[1] <- "cpg"  

identical(row.names(mval_mat), results_ordered_df$cpg %>% as.character())

results_ordered_df$fdr <- p.adjust(
    results_ordered_df$pValue,
    method = "fdr"
)

write.csv(
  results_ordered_df,
  paste0(data.dir.single.cpg.pval, "London_PFC_single_cpg_pVal_df.csv"),
  row.names = FALSE
)

```

```{R}
results_ordered_df <- readr::read_csv(paste0(data.dir.single.cpg.pval, "London_PFC_single_cpg_pVal_df.csv"))
results_ordered_df
```




# Linear regression by regions median Methylation 

## Residuals control and coMethylated Regions


1. Take residuals
2. Find co-methylated regions

Input: 

- QNBMIQ_PCfiltered
- pheno_withNeuronProp_df

Output: 

- QNBMIQ_PCfiltered_mvalResiduals
- residuals_cometh_ls

### Take residuals

```{R, eval = FALSE}
##### 1. Import datasets #######################################################
beta_mat <- readRDS(grep("QNBMIQ",dir(data.dir.pca,full.names = T),ignore.case = T,value = T)) 
pheno_df <- readRDS(dir(data.dir.neuron,full.names = T)) 

### Compute M values
mvalue_mat <- log2(beta_mat / (1 - beta_mat))

### Reorder samples based on pheno_df
mvalue_mat <- mvalue_mat[, pheno_df$sample]

identical(colnames(mvalue_mat),  pheno_df$sample)

### Take residuals
lmF <- function(mval){
  fitE <- lm(
    as.numeric(mval) ~ age.brain + sex + prop.neuron + as.character(slide), #add batch if rosmap
    data = pheno_df,
    na.action = na.exclude
  )
  residuals (fitE)
}
doParallel::registerDoParallel(cores = 16)
resid <- plyr::adply(mvalue_mat,1,.fun = lmF,.progress = "time",.parallel = TRUE)
rownames(resid) <- resid[,1]
resid[,1] <- NULL
colnames(resid) <- colnames(mvalue_mat)
dim(resid)
# [1] 433656     22

dim(mvalue_mat)
# [1] 433656     22

### Save dataset
saveRDS(
  resid,
  paste0(data.dir.residuals,"London_PFC_QNBMIQ_PCfiltered_mvalResiduals.RDS")
)


##### 4. Find co-methylated regions ############################################

### Import datasets
mvalue_residuals_mat <- readRDS(
  paste0(data.dir.residuals, "London_PFC_QNBMIQ_PCfiltered_mvalResiduals.RDS")
)

### Call in functions
library(coMethDMR)
library(BiocParallel)

probes.cluster.all <- coMethDMR::getPredefinedCluster(arrayType = "450k",
                                                      clusterType = "regions")

ncores <- 10
### Find co-methylated clusters
coMeth_ls <- CoMethAllRegions(
  dnam = mvalue_residuals_mat,      
  betaToM = FALSE,                   
  CpGs_ls = probes.cluster.all,
  arrayType = "450k",
  rDropThresh_num = 0.4,
  minPairwiseCorr = NULL,
  method = "spearman",             
  returnAllCpGs = TRUE,              
  output = "all",
  nCores_int = ncores,
  progressbar = FALSE
)


saveRDS(
  coMeth_ls,
  paste0(data.dir.residuals,"London_PFC_residuals_cometh_input_ls.RDS")
)
```


## Linear regression by regions median Methylation 

1. Calculate medians by cluster and sample
2. linear regression

Input: 

- QNBMIQ_PCfiltered,
- pheno_withNeuronProp_df
- residuals_cometh_input_ls

Output: 

- info_df
- mediansMval_df
- linear_df_aging

### Calculate medians by cluster and sample

```{R, eval = FALSE}
cohort <- "London"

### Import datasets
beta_mat <- beta_mat <- readRDS(dir(data.dir.pca, pattern = "QNBMIQ", full.names = TRUE))
pheno_df <- readRDS(dir(data.dir.neuron,full.names = T)) 
mval_mat <- log2(beta_mat / (1 - beta_mat)) %>% as.matrix()
coMeth_ls <- readRDS(
  paste0(data.dir.residuals, "London_PFC_residuals_cometh_input_ls.RDS")
)

### Create info dataset
input_cometh <- data.frame(
  inputRegion = coMeth_ls$inputRegion_chr,
  nCoMethRegion = coMeth_ls$nCoMethRegions_num,
  coMethRegion = names(coMeth_ls$coMeth_ls),
  nCpGs = unlist(lapply(coMeth_ls$coMeth_ls, length), use.names = FALSE),
  stringsAsFactors = FALSE
)

input_cometh_nodup <- input_cometh[
  !duplicated(input_cometh$coMethRegion),
]
colnames(input_cometh_nodup) <- c(
  paste0(cohort, "_inputRegion"),
  paste0(cohort, "_nCoMethRegion"),
  paste0(cohort, "_coMethRegion"),
  paste0(cohort, "_nCpGs")
)

saveRDS(
  input_cometh_nodup,
  paste0(data.dir.median,  cohort, "_info_df.rds")
)

### Take median of probes in each cluster for each sample
filename <-  paste0(paste0(data.dir.median, cohort, "_mediansMval_df.rds"))

library(robustbase)
mval_mat <- mval_mat[rownames(mval_mat) %in% unlist(coMeth_ls$coMeth_ls),]
if(!file.exists(filename)){
  medianMval.df <- plyr::ldply(
    coMeth_ls$coMeth_ls[!duplicated(names(coMeth_ls$coMeth_ls))],
    function(probes){
      colMedians(mval_mat[as.character(probes),], na.rm = TRUE)
    },
    .progress = "time"
  )
  medianMval.df$.id <- NULL
  colnames(medianMval.df) <- colnames(mval_mat)
  saveRDS(medianMval.df, file = filename)
} else {
  medianMval.df <- readRDS(filename)
}
```

### Test all regions -- linear regressions

```{R, eval = TRUE}
### Import datasets
cohort <- "London"
info_df <- readRDS(dir(data.dir.median,pattern = "info", full.names = TRUE))
mediansMval_df <- readRDS(dir(data.dir.median,pattern = "mediansMval", full.names = TRUE))
pheno_df <- readRDS(dir(data.dir.neuron,pattern = "NeuronProp_df", full.names = TRUE))

### Check variables before fitting model
pheno_df$Sample <- pheno_df$sample

identical(pheno_df$Sample, colnames(mediansMval_df))

pheno_df$sex <- as.factor(pheno_df$sex)
pheno_df$slide <- as.factor(pheno_df$slide)

str(pheno_df)
```

```{R, eval = FALSE}
# If rosmap cohort, don't forget batch effect
predictors_char <- "stage"
covariates_char <- c("age.brain", "sex", "prop.neuron", "slide")

devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")

res_df <- TestAllRegions_noInfo(
  predictors_char = predictors_char,
  covariates_char = covariates_char,
  pheno_df = pheno_df,
  summarizedRegions_df = mediansMval_df
)

colnames(res_df) <- c(
  paste0(cohort, "_estimate"),
  paste0(cohort, "_se"),
  paste0(cohort, "_pVal"),
  paste0(cohort, "_fdr")
)

res_withInfo_df <- cbind(info_df, res_df)

saveRDS(
  res_withInfo_df,
  paste0(data.dir.median, cohort, "_linear_df.rds")
)
```

```{R}
file <- dir(data.dir.median,pattern = paste0(".*linear_df"),
            recursive = T,
            full.names = TRUE,
            ignore.case = T)
file
res_withInfo_df <- readRDS(file)
dim(res_withInfo_df)
res_withInfo_df
```


# Data final

```{R}
dir(path = data.dir,recursive = T,pattern = ".rda|.csv|.RDS")
```

# Session information
```{R}
devtools::session_info()
```
