---
title: "GASPARONI dataset"
author: "Lanyu Zhang, Tiago C. Silva, Lily Wang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    theme: lumen
    highlight: kate
    toc: true
    number_sections: true
    df_print: paged
    code_download: true
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data retrival

```{R, message = FALSE, results = 'hide'}
library(dplyr)
library(minfi)
library(IlluminaHumanMethylation450kmanifest)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```

```{R}
cohort <- "GASPARONI"
data.dir <- file.path("../DATASETS",cohort) 
data.dir.table <- "../DATASETS/Summary_Table/" 
data.dir.raw <- "../../coMethDMR_metaAnalysis/code_validation/Meta_analysis_code/DATASETS/GASPARONI/step2_read_minfi/"
data.dir.bsfilter <- file.path(data.dir,"step2_bsConvFilter/") 
data.dir.clinical.filter <- file.path(data.dir,"step3_clinical_available_filtering/") 
data.dir.probes.qc <- file.path(data.dir,"step4_probesQC_filtering/") 
data.dir.probes.normalization <- file.path(data.dir,"step5_normalization/") 
data.dir.pca <- file.path(data.dir,"step6_pca_filtering/") 
data.dir.neuron <- file.path(data.dir,"step7_neuron_comp/") 
data.dir.single.cpg.pval <- file.path(data.dir,"step8_single_cpg_pval/") 
data.dir.residuals <- file.path(data.dir,"step9_residuals/") 
data.dir.median <- file.path(data.dir,"step10_median/") 
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Confirm sex status sing DNA methylation values

```{R}
load(file = paste0(data.dir.raw, "Gasparoni.rda"))

# Create a MethylSet object from RGSet
MSet <- preprocessRaw(RGSet)

# Create [Genomic]MethylSet object
GMset <- mapToGenome(MSet)

# Get predicted sex status
estSex <- getSex(GMset)

# Compare predicted gender with phenotype gender
realSex <- data.frame(
  sample = paste(phenoData$geo_accession, 
                 phenoData$sentrix_id.ch1, 
                 phenoData$sentrix_position.ch1, 
                 sep = "_"),
  realSex = phenoData$Sex.ch1,
  stringsAsFactors = FALSE
)
compareSex <- merge(
  as.data.frame(estSex), realSex,
  by.x = "row.names",
  by.y = "sample")

identical(compareSex$predictedSex, compareSex$realSex)

print(compareSex)
```

# Data QC 

## Bilsufite conversion filtering

Removing samples with bisulfiteConversion lower than 88.

```{R qc_samples_packages, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
```

```{R}
load(file = paste0(data.dir.raw, "Gasparoni.rda"))
```

```{R, include = FALSE}
aux <- phenoData  %>% as.data.frame() %>%
  dplyr::filter(
    source_name_ch1 == "Frontal Cortex" &
      !is.na(phenoData$braak_stage.ch1) &
      phenoData$characteristics_ch1 == "cell type: bulk"
  )
aux <- aux[match(substr(colnames(RGSet),1,10), aux$geo_accession),]
nb.female.samples <- sum(aux$Sex.ch1 == "F")
nb.male.samples <- sum(aux$Sex.ch1 == "M")
```

```{R}
bs <- data.frame(bisulfiteConversion = bscon(RGSet))
bsFilteredOut <- row.names(bs)[bs$bisulfiteConversion < 88]
nb.samples <- ncol(RGSet)
RGSet <- RGSet[,!colnames(RGSet) %in% bsFilteredOut]
nb.samples.bc.filtered <-  ncol(RGSet)
```

```{R}
phenoData <- phenoData[match(substr(colnames(RGSet),1,10), phenoData$geo_accession),]
nb.female.samples.bc.filtered <- sum(phenoData$Sex.ch1 == "F")
nb.male.samples.bc.filtered <- sum(phenoData$Sex.ch1 == "M")
```

```{R, eval = FALSE, include = FALSE}
save(RGSet,
     nb.samples,
     bs,
     phenoData,
     nb.samples.bc.filtered, 
     file = paste0(data.dir.bsfilter, "/RGSet_bsfiltered.rda"))
```

```{R, include = FALSE, eval = TRUE}
load(file = paste0(data.dir.bsfilter, "/RGSet_bsfiltered.rda"))
```

```{R, include = FALSE, eval=FALSE}
ggpubr::gghistogram(bs$bisulfiteConversion,xlab = "bisulfite Conversion" )
```

## Clinical data filtering

```{R}
dim(phenoData)

phenoData$braak_stage.ch1 <- phenoData$braak_stage.ch1 %>% as.numeric()
phenoData$age.ch1 <- phenoData$age.ch1 %>% as.numeric()
### Subset rows and columns
pheno_df <- phenoData  %>% as.data.frame() %>%
  dplyr::filter(
    source_name_ch1 == "Frontal Cortex" &
      !is.na(phenoData$braak_stage.ch1) &
      phenoData$characteristics_ch1 == "cell type: bulk"
  ) %>% dplyr::select(
    c(
      "geo_accession",
      "donor_id.ch1",
      "sentrix_id.ch1",
      "age.ch1",
      "Sex.ch1",
      "braak_stage.ch1"
    )
  )

### Rename vars
colnames(pheno_df) <- c(
  "sample", "subject.id", "slide", "age.brain", "sex", "stage"
)

dim(pheno_df) 
```

```{R, include = FALSE}
nb.samples.with.clinical <- nrow(pheno_df)
nb.male.samples.with.clinical <- sum(pheno_df$sex == "M")
nb.female.samples.with.clinical  <- sum(pheno_df$sex == "F")

## phenotype dataset
save(RGSet,
     nb.samples.with.clinical,
     pheno_df,
     file = paste0(data.dir.clinical.filter, "/gasparoni_bs_and_clinical_filtered.rda"))
```

## Probes QC

Input: 

- RGSet.RDS
- beta_mat.RDS

Output: 

- beta_CG_XY_SNPfiltered.RDS


```{R}
##### 1. subset to probes with detection P <= 0.01 #############################
### Read in RGSet and beta_mat
load(paste0(data.dir.clinical.filter, "/gasparoni_bs_and_clinical_filtered.rda"))
```

```{R, message = FALSE, results = 'hide'}
### subset to probes with detection P <= 0.01
library(minfi)
```

```{R}
beta_mat <- getBeta(RGSet) 
nb.probes <- nrow(beta_mat)
detP <- detectionP(RGSet, type = "m+u")
failed.01 <- detP > 0.01
passedProbes <- rownames(failed.01)[rowMeans(failed.01) == 0] 

beta_mat <- beta_mat[passedProbes, ]
nb.probes.detectP <- nrow(beta_mat)

##### 2. keep only probes that start with "cg" #################################
beta_mat <- beta_mat[grep("cg",rownames(beta_mat)),]
dim(beta_mat) 
nb.probes.detectP.cg <- nrow(beta_mat)

##### 3. drop probes where SNP with MAF >= 0.01 in the last 5 bp of the probe ##
```

```{R, message = FALSE, warning = FALSE}
library(DMRcate)
beta_mat <- rmSNPandCH(
  object = beta_mat,
  dist = 5, 
  mafcut = 0.01, 
  and = TRUE,
  rmcrosshyb = FALSE,
  rmXY = FALSE
) 
nb.probes.cg.dmrcate <- nrow(beta_mat)

##### 4. drop probes in chrM ###################################################
library(sesame)

probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

beta_mat <- beta_mat[
  row.names(beta_mat) %in% row.names(probes.info)[!probes.info$seqnames %in% "chrM"],
  ]
nb.probes.cg.dmrcate.chrM <- nrow(beta_mat)

##### 5. Output datasets #######################################################
save(
  pheno_df,
  nb.probes.detectP,
  nb.probes.detectP.cg,
  nb.probes.cg.dmrcate,
  nb.probes.cg.dmrcate.chrM,
  beta_mat,
  file = paste0(data.dir.probes.qc, "beta_CG_XY_SNPfiltered_mat.rda")
)
```

# Normalization

- Quantile normalization and BMIQ normalization

Input: 

- beta_CG_XY_SNPfiltered_mat.RDS
- RGSet.RDS
- pheno_df.RDS
- full.annot.RDS

Output: 

- bs.csv
- pheno_df.RDS
- QNBMIQ.RDS

```{R qcSamplesData}
load(paste0(data.dir.probes.qc, "beta_CG_XY_SNPfiltered_mat.rda"))
```


## Quantile normalization

### Check if sex chrom have higher/lower intensities than autosomes

#### Create function to extract separate matrix for each chromosome

```{R, message = FALSE, results = 'hide'}
library(sesame)
library(ggplot2)
```

```{R}
probes.info <- sesameDataGet("HM450.hg19.manifest")
probes.info <- probes.info[row.names(beta_mat) %>% as.character()] %>%
  as.data.frame %>%
  dplyr::select(c("seqnames","start","end"))
probes.info$seqnames <- as.character(probes.info$seqnames)

findBetaChr <- function(data, chrom){
  data %>%
    as.data.frame() %>% # has to turn matrix into df for next step
    tibble::rownames_to_column() %>% # turn rownames to a column, so rownames won't be deleted after filtering rows in next step
    filter(rowname %in% row.names(probes.info[probes.info$seqnames == chrom,])) %>%
    tibble::column_to_rownames() %>%
    as.matrix()
}
```

#### Compare beta values between gender

##### on chrX

```{R}
betaChrX <- findBetaChr(data = beta_mat, chrom = "chrX")
betaChrX_long <- data.frame(
  beta = as.vector(betaChrX),
  sample = rep(substr(colnames(betaChrX), 1,10), each = nrow(betaChrX)),
  stringsAsFactors = FALSE
)
betaChrX_long <- merge(
  betaChrX_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)

ggplot(betaChrX_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start = 1, end = 0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome X") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

##### on chrY

```{R}
betaChrY <- findBetaChr(data = beta_mat, chrom = "chrY")
betaChrY_long <- data.frame(
  beta = as.vector(betaChrY),
  sample = rep(substr(colnames(betaChrY), 1,10), each = nrow(betaChrY)),
  stringsAsFactors = FALSE
)
betaChrY_long <- merge(
  betaChrY_long, pheno_df[, c("sample", "sex")],
  by = "sample",
  sort = FALSE
)

ggplot(betaChrY_long,
       aes(x = sample, y = beta, fill = sex)) +
  stat_boxplot(geom ='errorbar', width = 1, linetype = 1) +
  geom_boxplot(width = 1, alpha = 1, outlier.shape = 1, outlier.size = 2) +
  scale_fill_grey(start=1, end=0.6) +
  labs(x = "sex", y = "DNA methylation beta values",
       title = "DNA methylation level by sex on chromosome Y") +
  theme_bw() +
  theme(legend.position = "none", axis.text.x = element_blank()) +
  facet_wrap(~sex)
```

#### Boxplot of methylation by chromosomes  

##### overall  

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)

chrSex <- paste0("chr", c("X", "Y"))
betaChrSex_ls <- lapply(seq_along(chrSex), function(i){
  dat <- findBetaChr(data = beta_mat, chrom = chrSex[i])
  data.frame(beta = as.vector(dat), chrom = chrSex[i], stringsAsFactors = FALSE)
})
betaChrSex_df <- do.call(rbind, betaChrSex_ls)

betaChr_df <- rbind(betaChrAuto_df, betaChrSex_df)

# # No enough memory to load all the data points and plot the figure, so use function boxplot instead
# ggplot(betaChr_df,
#        aes(x = chrom, y = beta, fill = chrom)) +
#   stat_boxplot(geom ='errorbar', width = 0.4, linetype = 1) +
#   geom_boxplot(width = 0.4, alpha = 1, outlier.shape = 1, outlier.size = 2) +
#   scale_fill_grey(start=1, end=0.6) +
#   labs(x = "chromosome type", y = "DNA methylation beta values",
#        title = "DNA methylation level vs. chromosome type") +
#   theme_bw()

boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")

pdf(file = paste0(data.dir.probes.normalization, "beta_vs_chrom.pdf"))
boxplot(
  beta ~ chrom, data = betaChr_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes")
dev.off()
```

##### by gender  

```{R}
### split matrix by sex first
beta_mat_female <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "F"]]
beta_mat_male <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "M"]]

### for females
chrAuto <- paste0("chr", 1:22)
female_auto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat_female, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
female_auto_df <- do.call(rbind, female_auto_ls)

female_sex_df <- findBetaChr(data = beta_mat_female, chrom = "chrX")
female_sex_df <- data.frame(beta = as.vector(female_sex_df), chrom = "chromosome X", stringsAsFactors = FALSE)

female_df <- rbind(female_auto_df, female_sex_df)

boxplot(
  beta ~ chrom, 
  data = female_df, 
  ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for females"
)

### for males
chrAuto <- paste0("chr", 1:22)
male_auto_ls <- lapply(seq_along(chrAuto), function(i){
  dat <- findBetaChr(data = beta_mat_male, chrom = chrAuto[i])
  data.frame(beta = as.vector(dat), chrom = "autosomes", stringsAsFactors = FALSE)
})
male_auto_df <- do.call(rbind, male_auto_ls)

male_X_df <- findBetaChr(data = beta_mat_male, chrom = "chrX")
male_X_df <- data.frame(beta = as.vector(male_X_df), chrom = "chromosome X", stringsAsFactors = FALSE)

male_Y_df <- findBetaChr(data = beta_mat_male, chrom = "chrY")
male_Y_df <- data.frame(beta = as.vector(male_Y_df), chrom = "chromosome Y", stringsAsFactors = FALSE)

male_df <- rbind(male_auto_df, male_X_df, male_Y_df)

boxplot(
  beta ~ chrom, data = male_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for males")

boxplot(
  beta ~ chrom, data = female_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for females")

### save plots
pdf(file = paste0(data.dir.probes.normalization, "beta_vs_chrom_by_gender.pdf"))

boxplot(
  beta ~ chrom, data = female_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for females")

boxplot(
  beta ~ chrom, data = male_df, ylab = "DNA methylation beta values",
  main = "DNA methylation on chromosomes for males")

dev.off()
```

#### Boxplot of methylation by gender on autosomes, X, Y

```{R, message = FALSE, results = 'hide'}
library(quantro)
```

##### on autosomes

```{R}
chrAuto <- paste0("chr", 1:22)
betaChrAuto_ls <- lapply(seq_along(chrAuto), function(i){findBetaChr(data = beta_mat, chrom = chrAuto[i])})
betaChrAuto_df <- do.call(rbind, betaChrAuto_ls)
matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrX 

```{R}
betaChrX_df <- findBetaChr(data = beta_mat, chrom = "chrX")
matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### on chrY

```{R}
betaChrY_df <- findBetaChr(data = beta_mat,chrom = "chrY")
matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

##### save plots

```{R}
#pdf (
#  paste0(data.dir.probes.normalization, "boxPlotBeforeNormalization.pdf")
#)

matboxplot(betaChrAuto_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrX_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaChrY_df,
           groupFactor = pheno_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - before normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

#dev.off()
```

### normalization by group

```{R, message = FALSE, results = 'hide'}
library(lumi)
```

#### split beta_mat into sub-matrics by females autosomes, females X, males autosomes, males X, males Y

```{R}
### split matrix by sex first
beta_mat_female <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "F"]]
beta_mat_male <- beta_mat[ 
  ,substr(colnames(beta_mat), 1, 10) %in% pheno_df$sample[pheno_df$sex == "M"]]

### split female beta matrix by chromosome (auto, X)
chrAuto <- paste0("chr", 1:22)
beta_mat_female_auto_ls <- lapply(seq_along(chrAuto), function(i){
  findBetaChr(data = beta_mat_female, chrom = chrAuto[i])})
beta_mat_female_auto <- do.call(rbind, beta_mat_female_auto_ls)

beta_mat_female_X <- findBetaChr(data = beta_mat_female, chrom = "chrX")

### split male beta matrix by chromosome (auto, X, Y)
chrAuto <- paste0("chr", 1:22)
beta_mat_male_auto_ls <- lapply(
  seq_along(chrAuto), 
  function(i){
    findBetaChr(data = beta_mat_male, chrom = chrAuto[i])
  })

beta_mat_male_auto <- do.call(rbind, beta_mat_male_auto_ls)
beta_mat_male_X <- findBetaChr(data = beta_mat_male, chrom = "chrX")
beta_mat_male_Y <- findBetaChr(data = beta_mat_male, chrom = "chrY")
```

#### normalization 

```{R}
### female autosomes
betaQN_female_auto <- lumiN(x.lumi = beta_mat_female_auto, method = "quantile")
dim(betaQN_female_auto)

### female chromosome X
betaQN_female_X <- lumiN(x.lumi = beta_mat_female_X, method = "quantile")
dim(betaQN_female_X)

### male autosomes
betaQN_male_auto <- lumiN(x.lumi = beta_mat_male_auto, method = "quantile")
dim(betaQN_male_auto)

### male chromosome X
betaQN_male_X <- lumiN(x.lumi = beta_mat_male_X, method = "quantile")
dim(betaQN_male_X)

### male chromosome Y
betaQN_male_Y <- lumiN(x.lumi = beta_mat_male_Y, method = "quantile")
dim(betaQN_male_Y)
```

##### Female plots after normalization 

```{R}
pheno_female_df <- pheno_df %>% filter(sex == "M")
matboxplot(betaQN_female_auto,
           groupFactor = pheno_female_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_female_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_female_X,
           groupFactor = pheno_female_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_female_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

```

##### Male plots after normalization 

```{R}
pheno_male_df <- pheno_df %>% filter(sex == "M")
matboxplot(betaQN_male_auto,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (autosomes) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_male_X,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome X) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)

matboxplot(betaQN_male_Y,
           groupFactor = pheno_male_df$sex,
           xaxt = "n",
           main = "Beta Values (chromosome Y) - after normalization")
legend('bottom',
       paste0("sex ", levels(as.factor(pheno_male_df$sex))),
       col = c(1:7), lty = 1, lwd = 3, cex = 0.70)
```

## BMIQ

```{R, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
library(sesame)
library(sesameData)
library(sm)
```

### test percentage of type I and type II probes then normalize the matrix

#### on autosomes

##### for females

```{R}
### Order annotation in the same order as beta matrix
annotType <- sesameDataGet("HM450.hg19.manifest")
annotType$designTypeNumeric <- ifelse(annotType$designType == "I",1,2)

### Density plot for type I and type II probes
betaQNCompleteCol1_female_auto <- betaQN_female_auto[complete.cases(betaQN_female_auto[,1]), ]
annotTypeCompleteCol1_female_auto <- annotType[row.names(betaQNCompleteCol1_female_auto), ]

sm.density.compare(
  betaQNCompleteCol1_female_auto[,1],
  annotTypeCompleteCol1_female_auto$designTypeNumeric) +
  title(main = "Density plot for type I and type II probes on female autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_auto),names(annotType))]
table(type12)
```

```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_female_auto <- plyr::aaply(
  betaQN_female_auto, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_female_auto) <- substr(colnames(betaQN_BMIQ_female_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_auto <- betaQN_male_auto[complete.cases(betaQN_male_auto[,1]), ]
annotTypeCompleteCol1_male_auto <- annotType[row.names(betaQNCompleteCol1_male_auto), ]

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_auto),names(annotType))]
table(type12)
```

```{R}
set.seed(946)
doParallel::registerDoParallel(cores = 4)
betaQN_BMIQ_male_auto <- plyr::aaply(
  betaQN_male_auto, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()
colnames(betaQN_BMIQ_male_auto) <- substr(colnames(betaQN_BMIQ_male_auto),1,stringr::str_length(pheno_df$sample) %>% unique)
```

#### on chromosome X

##### for females

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_female_X <- betaQN_female_X[complete.cases(betaQN_female_X[,1]), ]
annotTypeCompleteCol1_female_X <- annotType[row.names(betaQNCompleteCol1_female_X), ]

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_female_X),names(annotType))]
table(type12)
```

##### for males

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_X <- betaQN_male_X[complete.cases(betaQN_male_X[,1]), ]
annotTypeCompleteCol1_male_X <- annotType[row.names(betaQNCompleteCol1_male_X), ]

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_X),names(annotType))]
table(type12)
```

#### on chromosome Y

```{R}
### Density plot for type I and type II probes
betaQNCompleteCol1_male_Y <- betaQN_male_Y[complete.cases(betaQN_male_Y[,1]), ]
annotTypeCompleteCol1_male_Y <- annotType[row.names(betaQNCompleteCol1_male_Y), ]

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

### Summary table
type12 <- annotType$designTypeNumeric[match(rownames(betaQN_male_Y),names(annotType))]
table(type12)
```

### save density plots and normalized datasets

```{R}
### save plots
pdf(paste0(data.dir.probes.normalization, "densityPlotByProbeType.pdf"))

sm.density.compare(
  betaQNCompleteCol1_female_auto[,1],
  annotTypeCompleteCol1_female_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female autosomes")

sm.density.compare(
  betaQNCompleteCol1_male_auto[,1],
  annotTypeCompleteCol1_male_auto$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male autosomes")

sm.density.compare(
  betaQNCompleteCol1_female_X[,1],
  annotTypeCompleteCol1_female_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on female chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_X[,1],
  annotTypeCompleteCol1_male_X$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on male chromosome X")

sm.density.compare(
  betaQNCompleteCol1_male_Y[,1],
  annotTypeCompleteCol1_male_Y$designTypeNumeric
)
title(main = "Density plot for type I and type II probes on chromosome Y")

dev.off()

### combined normalized matrices
betaQN_BMIQ_female <- rbind(betaQN_BMIQ_female_auto, betaQN_female_X)
betaQN_BMIQ_male <- rbind(betaQN_BMIQ_male_auto, betaQN_male_X, betaQN_male_Y)
save(betaQN_BMIQ_female, betaQN_BMIQ_male, pheno_df, file = paste0(data.dir.probes.normalization, "GASPARONI_QNBMIQ.rda"))

```

# Outliers detection - PCA analysis

- Select most variable probes and perform PCA analysis

Input: 

- GASPARONI_QNBMIQ.rds  
- pheno_df.RDS  

Output: 

- GASPARONI_PCs_usingBetas.csv, 
- PCA plots
- GASPARONI_QNBMIQ_PCfiltered.RDS
- pheno_df.RDS

```{R}
# plotPCA and OrderDataBySd functions
devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")
```

```{R}
### merge in group info
## add center and scale
## compare M values vs. beta values
## with and without center / scale
load(paste0(data.dir.probes.normalization, "GASPARONI_QNBMIQ.rda"))
```

## for females  

```{R}
pheno_female_df <- pheno_df %>% filter(sex == "F")
pheno_female_df$stage3 <- ifelse(
  pheno_female_df$stage %in% c(0,1,2), "0-2",
  ifelse(pheno_female_df$stage %in% c(3,4), "3-4", "5-6")
)
betaQN_BMIQ_female <- betaQN_BMIQ_female[ , pheno_female_df$sample]

### transform to m values
mvalue_mat <- log2(betaQN_BMIQ_female / (1 - betaQN_BMIQ_female)) #dim: 457402 30

### order matrix by most variable probes on top
betaOrd_mat <- OrderDataBySd(betaQN_BMIQ_female) #dim: 457402 30
mOrd_mat <- OrderDataBySd(mvalue_mat)  #dim: 457402 30

identical(pheno_female_df$sample, colnames(betaOrd_mat))
identical(pheno_female_df$sample, colnames(mOrd_mat))

pca <- prcomp(t(betaOrd_mat[1:50000, ]),
              center = TRUE,
              scale = TRUE)

d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])

meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)

meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)

out3sdPC1_1 <- meanPC1 - 3 * sdPC1
out3sdPC1_2 <- meanPC1 + 3 * sdPC1

out3sdPC2_1 <- meanPC2 - 3 * sdPC2
out3sdPC2_2 <- meanPC2 + 3 * sdPC2

d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1

d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1

readr::write_csv(d, paste0(data.dir.pca, "GASPARONI_PCs_usingBetas_female.csv"))
```

```{R, eval = TRUE, include = TRUE}
### 2.PCA plot
library(ggplot2)
library(ggrepel)

### Beata values
byStage <- plotPCA(
  dataset = "Gasparoni: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_female_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

### M values
byStage <- plotPCA(
  dataset = "Gasparoni: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_female_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```

```{R}
### Filter samples by PCA, save files
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered_female <- betaQN_BMIQ_female[, rownames(noOutliers)] #dim: 457402 29
saveRDS(betaQN_BMIQ_PCfiltered_female, paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered_female.RDS"))

pheno_female_df <- pheno_female_df[pheno_female_df$sample %in% rownames(noOutliers),] #dim: 29 7
saveRDS(pheno_female_df, paste0(data.dir.pca, "pheno_female_df.RDS"))
```

## for males  

```{R}
pheno_male_df <- pheno_df %>% filter(sex == "M")
pheno_male_df$stage3 <- ifelse(
  pheno_male_df$stage %in% c(0,1,2), "0-2",
  ifelse(pheno_male_df$stage %in% c(3,4), "3-4", "5-6")
)
betaQN_BMIQ_male <- betaQN_BMIQ_male[ , pheno_male_df$sample]

### transform to m values
mvalue_mat <- log2(betaQN_BMIQ_male / (1 - betaQN_BMIQ_male)) #dim: 457464 27

### order matrix by most variable probes on top
betaOrd_mat <- OrderDataBySd(betaQN_BMIQ_male) #dim: 457464 27
mOrd_mat <- OrderDataBySd(mvalue_mat)  #dim: 457464 27

identical(pheno_male_df$sample, colnames(betaOrd_mat))
identical(pheno_male_df$sample, colnames(mOrd_mat))

pca <- prcomp(t(betaOrd_mat[1:50000, ]),
              center = TRUE,
              scale = TRUE)

d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])

meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)

meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)

out3sdPC1_1 <- meanPC1 - 3 * sdPC1
out3sdPC1_2 <- meanPC1 + 3 * sdPC1

out3sdPC2_1 <- meanPC2 - 3 * sdPC2
out3sdPC2_2 <- meanPC2 + 3 * sdPC2

d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1

d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1

readr::write_csv(d, paste0(data.dir.pca, "GASPARONI_PCs_usingBetas_male.csv"))
```

```{R, eval = TRUE, include = TRUE}
### 2.PCA plot
library(ggplot2)
library(ggrepel)

### Beata values
byStage <- plotPCA(
  dataset = "Gasparoni: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_male_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)

### M values
byStage <- plotPCA(
  dataset = "Gasparoni: M values",
  expSorted_mat = mOrd_mat,
  pheno = pheno_male_df,
  group_char = "stage3",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```


```{R}
### Filter samples by PCA, save files
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered_male <- betaQN_BMIQ_male[, rownames(noOutliers)] #dim: 457464 27
saveRDS(betaQN_BMIQ_PCfiltered_male, paste0(data.dir.pca, "Gasparoni_QNBMIQ_PCfiltered_male.RDS"))

pheno_male_df <- pheno_male_df[pheno_male_df$sample %in% rownames(noOutliers),] #dim: 27 7
saveRDS(pheno_male_df, paste0(data.dir.pca, "pheno_male_df.RDS"))
```

# Summary after QC steps

## Data and metadata
```{R}
nb.samples.with.clinical.after.pca.female <- ncol(betaQN_BMIQ_PCfiltered_female)
nb.samples.with.clinical.after.pca.male <- ncol(betaQN_BMIQ_PCfiltered_male)
dim(betaQN_BMIQ_PCfiltered_female)
dim(betaQN_BMIQ_PCfiltered_male)
dim(pheno_female_df)
dim(pheno_male_df)
```

```{R warning=TRUE, show = FALSE}
pheno_female_df %>% 
  DT::datatable(filter = 'top',
                style = "bootstrap",
                extensions = 'Buttons',
                options = list(scrollX = TRUE, 
                               dom = 'Bfrtip',
                               buttons = I('colvis'),
                               keys = TRUE, 
                               pageLength = 10), 
                rownames = FALSE,
                caption = "Female samples metadata")

pheno_male_df %>% 
  DT::datatable(filter = 'top',
                style = "bootstrap",
                extensions = 'Buttons',
                options = list(scrollX = TRUE, 
                               dom = 'Bfrtip',
                               buttons = I('colvis'),
                               keys = TRUE, 
                               pageLength = 10), 
                rownames = FALSE,
                caption = "Male samples metadata")
```

```{R, eval = FALSE, include = FALSE}
ggpubr::gghistogram(data = pheno_female_df, x = "stage",bins = 8)
# ggpubr::gghistogram(data = pheno_female_df, x = "stage",bins = 8,facet.by = "sex",fill = "sex")
ggpubr::gghistogram(data = pheno_female_df, x = "age.brain",bins = 20)
# ggpubr::gghistogram(data = pheno_female_df, x = "age.brain",bins = 20,fill = "sex",facet.by = "sex")

ggpubr::gghistogram(data = pheno_male_df, x = "stage",bins = 8)
# ggpubr::gghistogram(data = pheno_male_df, x = "stage",bins = 8,facet.by = "sex",fill = "sex")
ggpubr::gghistogram(data = pheno_male_df, x = "age.brain",bins = 20)
# ggpubr::gghistogram(data = pheno_male_df, x = "age.brain",bins = 20,fill = "sex",facet.by = "sex")
```


## Numbers of samples and probes removed in each step

```{R}
df.samples <- data.frame(
  "Number of samples" =  c(nb.samples, 
                           nb.samples.bc.filtered,
                           nb.samples.with.clinical, 
                           nb.samples.with.clinical.after.pca.female + nb.samples.with.clinical.after.pca.male),
  "Description" = c("total number of samples",
                    "samples with bisulfate conversion > 88",
                    "samples with clinical data",
                    "Samples after PCA"),
  "Difference" = c("-",
                   nb.samples.bc.filtered - nb.samples ,
                   nb.samples.with.clinical - nb.samples.bc.filtered ,
                   nb.samples.with.clinical.after.pca.female + nb.samples.with.clinical.after.pca.male - nb.samples.with.clinical)
)    
df.samples           


df.female.samples <- data.frame(
  "Number of samples" =  c(nb.female.samples, 
                           nb.female.samples.bc.filtered,
                           nb.samples.with.clinical, 
                           nb.samples.with.clinical.after.pca.female),
  "Description" = c("Female total number of samples",
                    "Female samples with bisulfate conversion > 88",
                    "Female samples with clinical data",
                    "Female samples after PCA"),
  "Difference" = c("-",
                   nb.female.samples.bc.filtered - nb.female.samples ,
                   nb.female.samples.with.clinical - nb.female.samples.bc.filtered ,
                   nb.samples.with.clinical.after.pca.female - nb.female.samples.with.clinical)
)    
df.female.samples           


df.male.samples <- data.frame(
  "Number of samples" =  c(nb.male.samples, 
                           nb.male.samples.bc.filtered,
                           nb.male.samples.with.clinical, 
                           nb.samples.with.clinical.after.pca.male),
  "Description" = c("Male total number of samples",
                    "Male samples with bisulfate conversion > 88",
                    "Male samples with clinical data",
                    "Male samples after PCA"),
  "Difference" = c("-",
                   nb.male.samples.bc.filtered - nb.male.samples ,
                   nb.male.samples.with.clinical - nb.male.samples.bc.filtered ,
                   nb.samples.with.clinical.after.pca.male - nb.male.samples.with.clinical)
)    
df.male.samples
# Create summary table
df.probes <- data.frame(
  "Number of probes" = c(nb.probes,
                         nb.probes.detectP, 
                         nb.probes.detectP.cg,
                         nb.probes.cg.dmrcate,
                         nb.probes.cg.dmrcate.chrM),
  "Description" = c("total number of probes in raw data",
                    "detection P < 0.01",
                    "only probes that start with cg",
                    "DMRcate",
                    "after deleting probes on chrM"),
  "Difference" = c("-",
                   nb.probes.detectP - nb.probes ,
                   nb.probes.detectP.cg - nb.probes.detectP,
                   nb.probes.cg.dmrcate - nb.probes.detectP.cg,
                   nb.probes.cg.dmrcate.chrM - nb.probes.cg.dmrcate)
)
df.probes
```

```{R, eval = FALSE, include = FALSE}
save(df.samples,df.probes,file = file.path(data.dir.table, "GASPARONI_table.rda"))
```

# Session information

```{R}
devtools::session_info()
```